// @healthcatalyst/cashmere/esm5/lib/typeform-survey/typeform-survey.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/typeform-survey/typeform-survey.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * @return {?}
 */
export function throwErrorForMissingSurveyUri() {
    throw Error("SurveyUri must be specified on element hc-typeform-survey");
}
var TypeformWindow = /** @class */ (function (_super) {
    tslib_1.__extends(TypeformWindow, _super);
    function TypeformWindow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return TypeformWindow;
}(Window));
export { TypeformWindow };
if (false) {
    /** @type {?} */
    TypeformWindow.prototype.typeformEmbed;
}
var TypeformSurveyComponent = /** @class */ (function () {
    function TypeformSurveyComponent() {
        this._id = 'typef_orm_share';
    }
    Object.defineProperty(TypeformSurveyComponent.prototype, "surveyUri", {
        get: /**
         * @return {?}
         */
        function () {
            return this._surveyUri;
        },
        /**
         * TypeForm survey URI you want to use. Example: https://somecompany.typeform.com/to/surveyId?parameter=parametervalue
         */
        set: /**
         * TypeForm survey URI you want to use. Example: https://somecompany.typeform.com/to/surveyId?parameter=parametervalue
         * @param {?} uri
         * @return {?}
         */
        function (uri) {
            this._surveyUri = uri;
            this.refreshFullUri();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TypeformSurveyComponent.prototype, "appVersion", {
        get: /**
         * @return {?}
         */
        function () {
            return this._appVersion;
        },
        /**
         * App version which will be passed to the survey in a hidden field. Ensures you know what version the feedback is referencing.
         */
        set: /**
         * App version which will be passed to the survey in a hidden field. Ensures you know what version the feedback is referencing.
         * @param {?} version
         * @return {?}
         */
        function (version) {
            this._appVersion = version;
            this.refreshFullUri();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Opens the survey specified in the surveyUri
     */
    /**
     * Opens the survey specified in the surveyUri
     * @return {?}
     */
    TypeformSurveyComponent.prototype.open = /**
     * Opens the survey specified in the surveyUri
     * @return {?}
     */
    function () {
        if (!document.getElementById(this._id)) {
            this.getScripts();
        }
        else {
            ((/** @type {?} */ (window))).typeformEmbed.makePopup(this._fullUri, {
                mode: 'drawer_right',
                autoOpen: true,
                opacity: 100,
                autoClose: 0,
                hideScrollbars: true
            });
        }
    };
    /**
     * @private
     * @return {?}
     */
    TypeformSurveyComponent.prototype.refreshFullUri = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var varChar = this.surveyUri.includes('?') ? '&' : '?';
        this._fullUri = this.appVersion ? this.surveyUri + varChar + 'app_version=' + this.appVersion : this.surveyUri;
    };
    /**
     * @private
     * @return {?}
     */
    TypeformSurveyComponent.prototype.getScripts = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.surveyUri) {
            throwErrorForMissingSurveyUri();
        }
        /* this is directly from the embed
                   markup given from TypeForm */
        /** @type {?} */
        var embedScript;
        /** @type {?} */
        var firstScript;
        if (!document.getElementById(this._id)) {
            // create new embed script with typeform cdn source
            embedScript = document.createElement('script');
            embedScript.id = this._id;
            embedScript.src = "https://embed.typeform.com/embed.js";
            // insert embed script before other js scripts
            firstScript = document.getElementsByTagName.call(document, 'script')[0];
            if (firstScript.parentNode) {
                firstScript.parentNode.insertBefore(embedScript, firstScript);
            }
        }
    };
    TypeformSurveyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-typeform-survey',
                    template: "\n        <a\n            class=\"typeform-share link\"\n            [href]=\"_fullUri\"\n            data-mode=\"drawer_right\"\n            data-auto-open=\"true\"\n            data-submit-close-delay=\"0\"\n            target=\"_blank\"\n            rel=\"noopener\"\n        ></a>\n    ",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    TypeformSurveyComponent.propDecorators = {
        surveyUri: [{ type: Input }],
        appVersion: [{ type: Input }]
    };
    return TypeformSurveyComponent;
}());
export { TypeformSurveyComponent };
if (false) {
    /** @type {?} */
    TypeformSurveyComponent.prototype._fullUri;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._surveyUri;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._appVersion;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._id;
}
