// @healthcatalyst/cashmere/esm2015/lib/form-field/hc-form.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-field/hc-form.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ContentChildren, QueryList } from '@angular/core';
import { parseBooleanAttribute } from '../util';
import { HcFormFieldComponent } from './hc-form-field.component';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
/**
 * `hcForm` directive that allows settings to be applied to all included HcFormFields
 */
export class HcFormDirective {
    constructor() {
        this._tight = false;
        this.unsubscribe$ = new Subject();
    }
    /**
     * Set the tight parameter on all enclosed HcFormFields. *Defaults to `false`.*
     * @return {?}
     */
    get tight() {
        return this._tight;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set tight(value) {
        this._tight = parseBooleanAttribute(value);
        this._updateTightFields();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._updateTightFields();
        // Pass the tight setting to any FormFields added dynamically
        this._formFields.changes.pipe(takeUntil(this.unsubscribe$)).subscribe((/**
         * @return {?}
         */
        () => this._updateTightFields()));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
    /**
     * @return {?}
     */
    _updateTightFields() {
        if (this._formFields) {
            this._formFields.forEach((/**
             * @param {?} field
             * @return {?}
             */
            field => {
                field.tight = this._tight;
            }));
        }
    }
}
HcFormDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hcForm]'
            },] }
];
HcFormDirective.propDecorators = {
    _formFields: [{ type: ContentChildren, args: [HcFormFieldComponent,] }],
    tight: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcFormDirective.prototype._tight;
    /**
     * @type {?}
     * @private
     */
    HcFormDirective.prototype.unsubscribe$;
    /** @type {?} */
    HcFormDirective.prototype._formFields;
}
