// @healthcatalyst/cashmere/esm5/lib/picklist/pane/picklist-pane.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/picklist/pane/picklist-pane.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { from } from 'rxjs';
/**
 * @template T
 */
var /**
 * @template T
 */
FilterableSelectList = /** @class */ (function () {
    function FilterableSelectList(options) {
        if (options === void 0) { options = new Map(); }
        this.options = options;
        this.selectedOptions = new Map();
        this.lastClickedOption = null;
        this.filteredOptions = new Array();
        this.isActive = true;
        this.additionalRemoteOptions = 0;
        this.loadingOptions = from([]);
        this.appendingOptions = from([]);
        this.codeIsSignificant = false;
    }
    return FilterableSelectList;
}());
/**
 * @template T
 */
export { FilterableSelectList };
if (false) {
    /** @type {?} */
    FilterableSelectList.prototype.selectedOptions;
    /** @type {?} */
    FilterableSelectList.prototype.lastClickedOption;
    /** @type {?} */
    FilterableSelectList.prototype.filteredOptions;
    /** @type {?} */
    FilterableSelectList.prototype.isActive;
    /** @type {?} */
    FilterableSelectList.prototype.additionalRemoteOptions;
    /** @type {?} */
    FilterableSelectList.prototype.loadingOptions;
    /** @type {?} */
    FilterableSelectList.prototype.appendingOptions;
    /** @type {?} */
    FilterableSelectList.prototype.codeIsSignificant;
    /** @type {?} */
    FilterableSelectList.prototype.sort;
    /** @type {?} */
    FilterableSelectList.prototype.options;
}
var SubSelectList = /** @class */ (function (_super) {
    tslib_1.__extends(SubSelectList, _super);
    function SubSelectList(parentValueSet, options) {
        if (options === void 0) { options = new Map(); }
        var _this = _super.call(this, options) || this;
        _this.parentValueSet = parentValueSet;
        _this.options = options;
        return _this;
    }
    return SubSelectList;
}(FilterableSelectList));
export { SubSelectList };
if (false) {
    /** @type {?} */
    SubSelectList.prototype.parentValueSet;
    /** @type {?} */
    SubSelectList.prototype.options;
}
/**
 * @param {?} model
 * @return {?}
 */
export function isSubList(model) {
    /** @type {?} */
    var subList = (/** @type {?} */ (model));
    return subList && subList.parentValueSet !== undefined;
}
var SelectListOption = /** @class */ (function () {
    function SelectListOption(option, code) {
        this.option = option;
        this.code = code;
        this.selected = false;
    }
    return SelectListOption;
}());
export { SelectListOption };
if (false) {
    /** @type {?} */
    SelectListOption.prototype.selected;
    /** @type {?} */
    SelectListOption.prototype.option;
    /** @type {?} */
    SelectListOption.prototype.code;
}
var ValueListOption = /** @class */ (function (_super) {
    tslib_1.__extends(ValueListOption, _super);
    function ValueListOption(option, code) {
        return _super.call(this, option, code) || this;
    }
    return ValueListOption;
}(SelectListOption));
export { ValueListOption };
var ValueSetListOption = /** @class */ (function (_super) {
    tslib_1.__extends(ValueSetListOption, _super);
    function ValueSetListOption(option, code) {
        var _this = _super.call(this, option, code) || this;
        _this.showValues = false;
        _this.loadingValues = false;
        _this.subValuesSelectList = new SubSelectList(_this);
        return _this;
    }
    return ValueSetListOption;
}(SelectListOption));
export { ValueSetListOption };
if (false) {
    /** @type {?} */
    ValueSetListOption.prototype.subValuesSelectList;
    /** @type {?} */
    ValueSetListOption.prototype.showValues;
    /** @type {?} */
    ValueSetListOption.prototype.loadingValues;
}
var PicklistValueOptions = /** @class */ (function () {
    function PicklistValueOptions() {
        this.values = new Map();
        this.valueSets = new Map();
    }
    return PicklistValueOptions;
}());
export { PicklistValueOptions };
if (false) {
    /** @type {?} */
    PicklistValueOptions.prototype.values;
    /** @type {?} */
    PicklistValueOptions.prototype.valueSets;
}
