// @healthcatalyst/cashmere/esm5/lib/sort/sort-header-intl.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sort/sort-header-intl.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Injectable, SkipSelf, Optional } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * To modify the labels and text displayed, create a new instance of HcSortHeaderIntl and
 * include it in a custom provider.
 */
var HcSortHeaderIntl = /** @class */ (function () {
    function HcSortHeaderIntl() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        /**
         * ARIA label for the sorting button.
         */
        this.sortButtonLabel = (/**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            return "Change sorting for " + id;
        });
    }
    HcSortHeaderIntl.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ HcSortHeaderIntl.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function HcSortHeaderIntl_Factory() { return new HcSortHeaderIntl(); }, token: HcSortHeaderIntl, providedIn: "root" });
    return HcSortHeaderIntl;
}());
export { HcSortHeaderIntl };
if (false) {
    /**
     * Stream that emits whenever the labels here are changed. Use this to notify
     * components if the labels have changed after initialization.
     * @type {?}
     */
    HcSortHeaderIntl.prototype.changes;
    /**
     * ARIA label for the sorting button.
     * @type {?}
     */
    HcSortHeaderIntl.prototype.sortButtonLabel;
}
/**
 * \@docs-private
 * @param {?} parentIntl
 * @return {?}
 */
export function HC_SORT_HEADER_INTL_PROVIDER_FACTORY(parentIntl) {
    return parentIntl || new HcSortHeaderIntl();
}
/**
 * \@docs-private
 * @type {?}
 */
export var HC_SORT_HEADER_INTL_PROVIDER = {
    // If there is already an HcSortHeaderIntl available, use that. Otherwise, provide a new one.
    provide: HcSortHeaderIntl,
    deps: [[new Optional(), new SkipSelf(), HcSortHeaderIntl]],
    useFactory: HC_SORT_HEADER_INTL_PROVIDER_FACTORY
};
