// @healthcatalyst/cashmere/esm5/lib/pagination/load-more-pagination.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pagination/load-more-pagination.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { validateStyleInput } from '../button/button.component';
import { BasePaginationComponent } from './base-pagination';
/**
 * A simple "load more" pagination button.
 * \@inheritdoc
 *
 */
var LoadMorePaginationComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LoadMorePaginationComponent, _super);
    function LoadMorePaginationComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Text to show on the button. *Defaults to 'Load more'.*
         */
        _this.buttonText = 'Load more';
        _this._style = 'secondary';
        return _this;
    }
    Object.defineProperty(LoadMorePaginationComponent.prototype, "buttonStyle", {
        /** Sets style of button. Choose from: `'primary' | 'primary-alt' | 'destructive' | 'neutral' | 'secondary' | 'link' | 'link-inline'`.
         * *Defaults to `'secondary'`.*
         */
        get: /**
         * Sets style of button. Choose from: `'primary' | 'primary-alt' | 'destructive' | 'neutral' | 'secondary' | 'link' | 'link-inline'`.
         * *Defaults to `'secondary'`.*
         * @return {?}
         */
        function () {
            return this._style;
        },
        set: /**
         * @param {?} btnStyle
         * @return {?}
         */
        function (btnStyle) {
            validateStyleInput(btnStyle);
            this._style = btnStyle;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    LoadMorePaginationComponent.prototype._loadNextPage = /**
     * @return {?}
     */
    function () {
        if (this._isLastPage) {
            return;
        }
        this.pageNumber = this.pageNumber + 1;
    };
    LoadMorePaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-load-more-pagination',
                    template: "<button hc-button class=\"hc-page-load-more-button\" [buttonStyle]=\"buttonStyle\" (click)=\"_loadNextPage()\" [disabled]=\"_isLastPage\">\n    {{buttonText}}\n</button>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".hc-page-load-more-button{display:block;margin:20px auto;min-width:30%}"]
                }] }
    ];
    LoadMorePaginationComponent.propDecorators = {
        buttonText: [{ type: Input }],
        buttonStyle: [{ type: Input }]
    };
    return LoadMorePaginationComponent;
}(BasePaginationComponent));
export { LoadMorePaginationComponent };
if (false) {
    /**
     * Text to show on the button. *Defaults to 'Load more'.*
     * @type {?}
     */
    LoadMorePaginationComponent.prototype.buttonText;
    /**
     * @type {?}
     * @private
     */
    LoadMorePaginationComponent.prototype._style;
}
