// @healthcatalyst/cashmere/esm2015/lib/pagination/base-pagination.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pagination/base-pagination.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output, Component } from '@angular/core';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { Initializable } from '../shared/initializable';
/**
 * Base Pagination class for shared functionality
 *
 */
export class BasePaginationComponent extends Initializable {
    constructor() {
        super(...arguments);
        this._length = 0;
        this._pageNumber = 1;
        this._pageSize = BasePaginationComponent._DEFAULT_PAGE_SIZE;
        /**
         * Event emitted when the paginator changes the page size or page index.
         */
        this.page = new EventEmitter();
        /**
         * Emits the new page number when the page number changes.
         */
        this.pageNumberChange = new EventEmitter();
        /**
         * Emits the new page size when the page size changes.
         */
        this.pageSizeChange = new EventEmitter();
    }
    /**
     * The total number of items to be paged through
     * @return {?}
     */
    get length() {
        return this._length;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set length(value) {
        this._length = coerceNumberProperty(value);
    }
    /**
     * The currently displayed page. *Defaults to 1.*
     * @return {?}
     */
    get pageNumber() {
        return this._pageNumber;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set pageNumber(value) {
        /** @type {?} */
        const prevPageNumber = this._pageNumber;
        this._pageNumber = value;
        /** @type {?} */
        const sanitizedValue = this._sanitizePageNumber(value);
        if (sanitizedValue !== value) {
            setTimeout((/**
             * @return {?}
             */
            () => (this.pageNumber = sanitizedValue)));
        }
        else {
            this.pageNumberChange.emit(this.pageNumber);
            this._emitPageEvent(prevPageNumber);
        }
    }
    /**
     * Number of items to display on a page. *Defaults to 20.*
     * @return {?}
     */
    get pageSize() {
        return this._pageSize;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set pageSize(value) {
        this._pageSize = coerceNumberProperty(value);
        this._pageSizeUpdated();
        this.pageSizeChange.emit(this.pageSize);
        this._emitPageEvent(this.pageNumber);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._markInitialized();
    }
    /**
     * The computed total number of pages
     * @return {?}
     */
    get totalPages() {
        return Math.ceil(this._length / this._pageSize);
    }
    /**
     * @return {?}
     */
    get _isFirstPage() {
        return this._pageNumber === 1;
    }
    /**
     * @return {?}
     */
    get _isLastPage() {
        return !!(this.totalPages && this._pageNumber === this.totalPages);
    }
    /**
     * Changes the page size so that the first item displayed on the page will still be
     * displayed using the new page size.
     *
     * For example, if the page size is 10 and on the second page (items indexed 11-20) then
     * switching so that the page size is 5 will set the third page as the current page so
     * that the 11th item will still be displayed.
     * @param {?} pageSize
     * @return {?}
     */
    _changePageSize(pageSize) {
        // Current page needs to be updated to reflect the new page size. Navigate to the page
        // containing the previous page's first item.
        /** @type {?} */
        const startIndex = (this.pageNumber - 1) * this.pageSize;
        this.pageSize = pageSize;
        this.pageNumber = Math.ceil(startIndex / pageSize) + 1;
    }
    /**
     * @return {?}
     */
    _pageSizeUpdated() { }
    /**
     * @private
     * @param {?} pageNumber
     * @return {?}
     */
    _sanitizePageNumber(pageNumber) {
        /** @type {?} */
        const number = Math.max(coerceNumberProperty(pageNumber), 1);
        return number > this.totalPages ? this.totalPages : number;
    }
    /**
     * @private
     * @param {?} previousPageNumber
     * @return {?}
     */
    _emitPageEvent(previousPageNumber) {
        this.page.emit({
            previousPageNumber,
            pageNumber: this.pageNumber,
            pageSize: this.pageSize,
            length: this.length
        });
    }
}
BasePaginationComponent._DEFAULT_PAGE_SIZE = 20;
BasePaginationComponent.decorators = [
    { type: Component, args: [{
                template: ''
            }] }
];
BasePaginationComponent.propDecorators = {
    length: [{ type: Input }],
    pageNumber: [{ type: Input }],
    pageSize: [{ type: Input }],
    page: [{ type: Output }],
    pageNumberChange: [{ type: Output }],
    pageSizeChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    BasePaginationComponent._DEFAULT_PAGE_SIZE;
    /**
     * @type {?}
     * @private
     */
    BasePaginationComponent.prototype._length;
    /**
     * @type {?}
     * @private
     */
    BasePaginationComponent.prototype._pageNumber;
    /**
     * @type {?}
     * @private
     */
    BasePaginationComponent.prototype._pageSize;
    /**
     * Event emitted when the paginator changes the page size or page index.
     * @type {?}
     */
    BasePaginationComponent.prototype.page;
    /**
     * Emits the new page number when the page number changes.
     * @type {?}
     */
    BasePaginationComponent.prototype.pageNumberChange;
    /**
     * Emits the new page size when the page size changes.
     * @type {?}
     */
    BasePaginationComponent.prototype.pageSizeChange;
}
