// @healthcatalyst/cashmere/esm2015/lib/pagination/load-more-pagination.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pagination/load-more-pagination.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { validateStyleInput } from '../button/button.component';
import { BasePaginationComponent } from './base-pagination';
/**
 * A simple "load more" pagination button.
 * \@inheritdoc
 *
 */
export class LoadMorePaginationComponent extends BasePaginationComponent {
    constructor() {
        super(...arguments);
        /**
         * Text to show on the button. *Defaults to 'Load more'.*
         */
        this.buttonText = 'Load more';
        this._style = 'secondary';
    }
    /**
     * Sets style of button. Choose from: `'primary' | 'primary-alt' | 'destructive' | 'neutral' | 'secondary' | 'link' | 'link-inline'`.
     * *Defaults to `'secondary'`.*
     * @return {?}
     */
    get buttonStyle() {
        return this._style;
    }
    /**
     * @param {?} btnStyle
     * @return {?}
     */
    set buttonStyle(btnStyle) {
        validateStyleInput(btnStyle);
        this._style = btnStyle;
    }
    /**
     * @return {?}
     */
    _loadNextPage() {
        if (this._isLastPage) {
            return;
        }
        this.pageNumber = this.pageNumber + 1;
    }
}
LoadMorePaginationComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-load-more-pagination',
                template: "<button hc-button class=\"hc-page-load-more-button\" [buttonStyle]=\"buttonStyle\" (click)=\"_loadNextPage()\" [disabled]=\"_isLastPage\">\n    {{buttonText}}\n</button>",
                encapsulation: ViewEncapsulation.None,
                styles: [".hc-page-load-more-button{display:block;margin:20px auto;min-width:30%}"]
            }] }
];
LoadMorePaginationComponent.propDecorators = {
    buttonText: [{ type: Input }],
    buttonStyle: [{ type: Input }]
};
if (false) {
    /**
     * Text to show on the button. *Defaults to 'Load more'.*
     * @type {?}
     */
    LoadMorePaginationComponent.prototype.buttonText;
    /**
     * @type {?}
     * @private
     */
    LoadMorePaginationComponent.prototype._style;
}
