// @healthcatalyst/cashmere/esm5/lib/pipes/number-abbreviator/number-abbreviator.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/number-abbreviator/number-abbreviator.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Pipe } from "@angular/core";
/**
 * Creates an abbreviated number for display. 1,234,234 => 1.23M
 */
var NumberAbbreviatorPipe = /** @class */ (function () {
    function NumberAbbreviatorPipe() {
    }
    /**
     * @param value the value to be abbreviated
     * @param decimalPoints how many decimal points to round to
     * @param threshold represents the number at which we should start abbreviating;
     * if the threshold is 1 million, we'll leave number less than 1 million as is (aside from adding commas or decimal points)
     */
    /**
     * @param {?} value the value to be abbreviated
     * @param {?=} decimalPoints how many decimal points to round to
     * @param {?=} threshold represents the number at which we should start abbreviating;
     * if the threshold is 1 million, we'll leave number less than 1 million as is (aside from adding commas or decimal points)
     * @return {?}
     */
    NumberAbbreviatorPipe.prototype.transform = /**
     * @param {?} value the value to be abbreviated
     * @param {?=} decimalPoints how many decimal points to round to
     * @param {?=} threshold represents the number at which we should start abbreviating;
     * if the threshold is 1 million, we'll leave number less than 1 million as is (aside from adding commas or decimal points)
     * @return {?}
     */
    function (value, decimalPoints, threshold) {
        var e_1, _a;
        if (decimalPoints === void 0) { decimalPoints = 2; }
        if (threshold === void 0) { threshold = 1000000; }
        if (value === null) {
            return value;
        }
        // use defaults if given funky parameters
        if (decimalPoints < 0 || !Number.isInteger(decimalPoints)) {
            decimalPoints = 2;
        }
        if (threshold < 1000 || !Number.isInteger(threshold)) {
            threshold = 1000;
        }
        /** @type {?} */
        var abs = Math.abs(value);
        /** @type {?} */
        var key = "";
        /** @type {?} */
        var roundedStringVal = "";
        /** @type {?} */
        var isNegative = value < 0;
        if (abs >= threshold) {
            try {
                // Abbreviate, adding unit if greater than specified limit
                for (var _b = tslib_1.__values(NumberAbbreviatorPipe.powers), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var power = _c.value;
                    /** @type {?} */
                    var reduced = abs / power.value;
                    reduced = this.roundValue(reduced, decimalPoints);
                    if (reduced >= 1) {
                        roundedStringVal = reduced.toString();
                        key = power.key;
                        break;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        else {
            // Just round decimals and add commas if less than specified limit
            roundedStringVal = this.roundValue(abs, decimalPoints).toLocaleString();
        }
        return (isNegative ? "-" : "") + roundedStringVal + key;
    };
    /**
     * @private
     * @param {?} value
     * @param {?=} decimalPoints
     * @return {?}
     */
    NumberAbbreviatorPipe.prototype.roundValue = /**
     * @private
     * @param {?} value
     * @param {?=} decimalPoints
     * @return {?}
     */
    function (value, decimalPoints) {
        if (decimalPoints === void 0) { decimalPoints = 2; }
        /** @type {?} */
        var rounder = Math.pow(10, decimalPoints);
        return Math.round(value * rounder) / rounder;
    };
    NumberAbbreviatorPipe.powers = [
        { key: "Q", value: Math.pow(10, 15) },
        { key: "T", value: Math.pow(10, 12) },
        { key: "B", value: Math.pow(10, 9) },
        { key: "M", value: Math.pow(10, 6) },
        { key: "K", value: 1000 }
    ];
    NumberAbbreviatorPipe.decorators = [
        { type: Pipe, args: [{ name: "abbreviateNumber" },] }
    ];
    return NumberAbbreviatorPipe;
}());
export { NumberAbbreviatorPipe };
if (false) {
    /** @type {?} */
    NumberAbbreviatorPipe.powers;
}
