// @healthcatalyst/cashmere/esm5/lib/pop/popover.errors.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pop/popover.errors.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { VALID_HORIZ_ALIGN, VALID_VERT_ALIGN, VALID_SCROLL, VALID_TRIGGER } from './types';
/**
 * @return {?}
 */
export function getInvalidPopoverError() {
    return Error('HcPopoverAnchorDirective must be provided an HcPopover component instance.');
}
/**
 * @return {?}
 */
export function getUnanchoredPopoverError() {
    return Error('HcPopover is not anchored to any HcPopoverAnchorDirective.');
}
/**
 * @param {?} alignment
 * @return {?}
 */
export function getInvalidHorizontalAlignError(alignment) {
    return Error(generateGenericError('horizontalAlign/xAlign', alignment, VALID_HORIZ_ALIGN));
}
/**
 * @param {?} alignment
 * @return {?}
 */
export function getInvalidVerticalAlignError(alignment) {
    return Error(generateGenericError('verticalAlign/yAlign', alignment, VALID_VERT_ALIGN));
}
/**
 * @param {?} trigger
 * @return {?}
 */
export function getInvalidTriggerError(trigger) {
    return Error(generateGenericError('trigger', trigger, VALID_TRIGGER));
}
/**
 * @param {?} strategy
 * @return {?}
 */
export function getInvalidScrollStrategyError(strategy) {
    return Error(generateGenericError('scrollStrategy', strategy, VALID_SCROLL));
}
/**
 * @param {?} apiName
 * @param {?} invalid
 * @param {?} valid
 * @return {?}
 */
function generateGenericError(apiName, invalid, valid) {
    return "Invalid " + apiName + ": '" + invalid + "'. Valid options are " + (valid.map((/**
     * @param {?} v
     * @return {?}
     */
    function (v) { return "'" + v + "'"; })).join(', ') + ".");
}
