// @healthcatalyst/cashmere/esm5/lib/pop/popover-accessibility.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pop/popover-accessibility.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/** @enum {number} */
var KEY_CODE = {
    DOWN_ARROW: 40,
    RIGHT_ARROW: 39,
    UP_ARROW: 38,
    LEFT_ARROW: 37,
    TAB: 9,
};
export { KEY_CODE };
KEY_CODE[KEY_CODE.DOWN_ARROW] = 'DOWN_ARROW';
KEY_CODE[KEY_CODE.RIGHT_ARROW] = 'RIGHT_ARROW';
KEY_CODE[KEY_CODE.UP_ARROW] = 'UP_ARROW';
KEY_CODE[KEY_CODE.LEFT_ARROW] = 'LEFT_ARROW';
KEY_CODE[KEY_CODE.TAB] = 'TAB';
/**
 * @record
 */
export function HcPopKeyboardNotifier() { }
if (false) {
    /** @type {?} */
    HcPopKeyboardNotifier.prototype.isOpen;
    /** @type {?} */
    HcPopKeyboardNotifier.prototype.nativeElement;
    /** @type {?} */
    HcPopKeyboardNotifier.prototype.hasSubmenu;
    /** @type {?} */
    HcPopKeyboardNotifier.prototype.onKeyDown;
}
var HcPopoverAccessibilityService = /** @class */ (function () {
    function HcPopoverAccessibilityService() {
        var _this = this;
        this.notifiers = [];
        this.handleKeydown = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.notifiers.filter((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.isOpen; })).forEach((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.onKeyDown(event); }));
            if (event.keyCode === KEY_CODE.RIGHT_ARROW) {
                _this.notifiers.filter((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return document.activeElement === n.nativeElement && n.hasSubmenu(); })).forEach((/**
                 * @param {?} n
                 * @return {?}
                 */
                function (n) { return n.onKeyDown(event); }));
            }
        });
        window.addEventListener('keydown', this.handleKeydown);
    }
    /**
     * @return {?}
     */
    HcPopoverAccessibilityService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        window.removeEventListener('keydown', this.handleKeydown);
    };
    /**
     * @param {?} notifier
     * @return {?}
     */
    HcPopoverAccessibilityService.prototype.registerNotifier = /**
     * @param {?} notifier
     * @return {?}
     */
    function (notifier) {
        this.notifiers.push(notifier);
    };
    HcPopoverAccessibilityService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HcPopoverAccessibilityService.ctorParameters = function () { return []; };
    return HcPopoverAccessibilityService;
}());
export { HcPopoverAccessibilityService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcPopoverAccessibilityService.prototype.notifiers;
    /**
     * @type {?}
     * @private
     */
    HcPopoverAccessibilityService.prototype.handleKeydown;
}
