// @healthcatalyst/cashmere/esm2015/lib/picklist/pane/picklist-pane.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/picklist/pane/picklist-pane.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { from } from 'rxjs';
/**
 * @template T
 */
export class FilterableSelectList {
    /**
     * @param {?=} options
     */
    constructor(options = new Map()) {
        this.options = options;
        this.selectedOptions = new Map();
        this.lastClickedOption = null;
        this.filteredOptions = new Array();
        this.isActive = true;
        this.additionalRemoteOptions = 0;
        this.loadingOptions = from([]);
        this.appendingOptions = from([]);
        this.codeIsSignificant = false;
    }
}
if (false) {
    /** @type {?} */
    FilterableSelectList.prototype.selectedOptions;
    /** @type {?} */
    FilterableSelectList.prototype.lastClickedOption;
    /** @type {?} */
    FilterableSelectList.prototype.filteredOptions;
    /** @type {?} */
    FilterableSelectList.prototype.isActive;
    /** @type {?} */
    FilterableSelectList.prototype.additionalRemoteOptions;
    /** @type {?} */
    FilterableSelectList.prototype.loadingOptions;
    /** @type {?} */
    FilterableSelectList.prototype.appendingOptions;
    /** @type {?} */
    FilterableSelectList.prototype.codeIsSignificant;
    /** @type {?} */
    FilterableSelectList.prototype.sort;
    /** @type {?} */
    FilterableSelectList.prototype.options;
}
export class SubSelectList extends FilterableSelectList {
    /**
     * @param {?} parentValueSet
     * @param {?=} options
     */
    constructor(parentValueSet, options = new Map()) {
        super(options);
        this.parentValueSet = parentValueSet;
        this.options = options;
    }
}
if (false) {
    /** @type {?} */
    SubSelectList.prototype.parentValueSet;
    /** @type {?} */
    SubSelectList.prototype.options;
}
/**
 * @param {?} model
 * @return {?}
 */
export function isSubList(model) {
    /** @type {?} */
    const subList = (/** @type {?} */ (model));
    return subList && subList.parentValueSet !== undefined;
}
export class SelectListOption {
    /**
     * @param {?} option
     * @param {?} code
     */
    constructor(option, code) {
        this.option = option;
        this.code = code;
        this.selected = false;
    }
}
if (false) {
    /** @type {?} */
    SelectListOption.prototype.selected;
    /** @type {?} */
    SelectListOption.prototype.option;
    /** @type {?} */
    SelectListOption.prototype.code;
}
export class ValueListOption extends SelectListOption {
    /**
     * @param {?} option
     * @param {?} code
     */
    constructor(option, code) {
        super(option, code);
    }
}
export class ValueSetListOption extends SelectListOption {
    /**
     * @param {?} option
     * @param {?} code
     */
    constructor(option, code) {
        super(option, code);
        this.showValues = false;
        this.loadingValues = false;
        this.subValuesSelectList = new SubSelectList(this);
    }
}
if (false) {
    /** @type {?} */
    ValueSetListOption.prototype.subValuesSelectList;
    /** @type {?} */
    ValueSetListOption.prototype.showValues;
    /** @type {?} */
    ValueSetListOption.prototype.loadingValues;
}
export class PicklistValueOptions {
    constructor() {
        this.values = new Map();
        this.valueSets = new Map();
    }
}
if (false) {
    /** @type {?} */
    PicklistValueOptions.prototype.values;
    /** @type {?} */
    PicklistValueOptions.prototype.valueSets;
}
