// @healthcatalyst/cashmere/esm5/lib/checkbox/checkbox-required.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/checkbox/checkbox-required.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/* tslint:disable:directive-selector */
import { Directive, forwardRef, HostBinding } from '@angular/core';
import { CheckboxRequiredValidator, NG_VALIDATORS } from '@angular/forms';
/** @type {?} */
export var HC_CHECKBOX_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return HcCheckboxRequiredValidatorDirective; })),
    multi: true
};
/**
 * \@docs-private
 */
var HcCheckboxRequiredValidatorDirective = /** @class */ (function (_super) {
    tslib_1.__extends(HcCheckboxRequiredValidatorDirective, _super);
    function HcCheckboxRequiredValidatorDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(HcCheckboxRequiredValidatorDirective.prototype, "isRequired", {
        get: /**
         * @return {?}
         */
        function () {
            return this.required ? '' : null;
        },
        enumerable: true,
        configurable: true
    });
    HcCheckboxRequiredValidatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: "hc-checkbox[required][formControlName],\n             hc-checkbox[required][formControl],hc-checkbox[required][ngModel]",
                    providers: [HC_CHECKBOX_REQUIRED_VALIDATOR]
                },] }
    ];
    HcCheckboxRequiredValidatorDirective.propDecorators = {
        isRequired: [{ type: HostBinding, args: ['attr.required',] }]
    };
    return HcCheckboxRequiredValidatorDirective;
}(CheckboxRequiredValidator));
export { HcCheckboxRequiredValidatorDirective };
