// @healthcatalyst/cashmere/esm5/lib/modal/modal.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ModalFooterComponent } from './modal-footer.component';
import { ModalWindowComponent } from './modal-window.component';
import { ModalOverlayComponent } from './modal-overlay.component';
import { ModalService } from './modal.service';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ModalHeaderComponent } from './modal-header.component';
import { ModalBodyComponent } from './modal-body.component';
import { ModalComponent } from './modal.component';
var ModalModule = /** @class */ (function () {
    function ModalModule() {
    }
    ModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [
                        ModalOverlayComponent,
                        ModalWindowComponent,
                        ModalHeaderComponent,
                        ModalBodyComponent,
                        ModalFooterComponent,
                        ModalComponent
                    ],
                    entryComponents: [ModalOverlayComponent, ModalWindowComponent],
                    exports: [ModalHeaderComponent, ModalBodyComponent, ModalFooterComponent, ModalComponent],
                    providers: [ModalService]
                },] }
    ];
    return ModalModule;
}());
export { ModalModule };
