// @healthcatalyst/cashmere/esm5/lib/modal/modal-window.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-window.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { animate, state, style, transition, trigger } from '@angular/animations';
import { Component, ElementRef, HostBinding, HostListener, Input, ViewEncapsulation } from '@angular/core';
import { ActiveModal } from './active-modal';
var ModalWindowComponent = /** @class */ (function () {
    function ModalWindowComponent(activeModal, el) {
        this.activeModal = activeModal;
        this.el = el;
        this._ignoreOverlayClick = false;
        this._size = 'auto';
    }
    /**
     * @return {?}
     */
    ModalWindowComponent.prototype._fadeInOut = /**
     * @return {?}
     */
    function () {
        return state;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ModalWindowComponent.prototype._overlayClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var e_1, _a;
        var _this = this;
        /** @type {?} */
        var modalContentNotPresent = true;
        /** @type {?} */
        var path = this._eventPath(event);
        /** @type {?} */
        var modalWindowTargetIncluded = path.findIndex((/**
         * @param {?} p
         * @return {?}
         */
        function (p) { return p === _this.el.nativeElement; })) > -1;
        /** @type {?} */
        var classList = path.map((/**
         * @param {?} p
         * @return {?}
         */
        function (p) { return p.classList; }));
        try {
            for (var classList_1 = tslib_1.__values(classList), classList_1_1 = classList_1.next(); !classList_1_1.done; classList_1_1 = classList_1.next()) {
                var cl = classList_1_1.value;
                if (cl) {
                    if (cl.contains('hc-modal-content')) {
                        modalContentNotPresent = false;
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (classList_1_1 && !classList_1_1.done && (_a = classList_1.return)) _a.call(classList_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        /* The hc-modal has 100% height (although not completely visible).
            To enable closing the modal while clicking above or below the modal (in
            addition to the sides) this function will check to see if the click
            event includes:
                1. This window element
                2. Not the hc-modal-content and
                3. the overlay click option is disabled. */
        if (!this._ignoreOverlayClick && modalContentNotPresent && modalWindowTargetIncluded) {
            this.activeModal.dismiss();
        }
    };
    // Serves as a polyfill for Event.composedPath() or Event.Path
    // Serves as a polyfill for Event.composedPath() or Event.Path
    /**
     * @param {?} evt
     * @return {?}
     */
    ModalWindowComponent.prototype._eventPath = 
    // Serves as a polyfill for Event.composedPath() or Event.Path
    /**
     * @param {?} evt
     * @return {?}
     */
    function (evt) {
        /** @type {?} */
        var path = (evt.composedPath && evt.composedPath()) || evt.path;
        /** @type {?} */
        var target = evt.target;
        if (path != null) {
            // Safari doesn't include Window, but it should.
            return path.indexOf(window) < 0 ? path.concat(window) : path;
        }
        if (target === window) {
            return [window];
        }
        /**
         * @param {?} node
         * @param {?=} memo
         * @return {?}
         */
        function _getParents(node, memo) {
            memo = memo || [];
            /** @type {?} */
            var parentNode = node.parentNode;
            if (!parentNode) {
                return memo;
            }
            else {
                return _getParents(parentNode, memo.concat(parentNode));
            }
        }
        return [target].concat(_getParents(target), window);
    };
    ModalWindowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-modal-window',
                    template: "\n        <div [class]=\"'hc-modal hc-modal-' + _size\"><ng-content></ng-content></div>\n    ",
                    encapsulation: ViewEncapsulation.None,
                    // tslint:disable-next-line: no-host-metadata-property
                    host: { class: 'hc-modal-window' },
                    animations: [
                        trigger('fadeInOut', [
                            state('in', style({ opacity: 1 })),
                            transition('void <=> *', [
                                style({
                                    opacity: 0
                                }),
                                animate('0.2s ease-in-out')
                            ])
                        ])
                    ],
                    styles: [".hc-modal-window{-webkit-box-align:center;align-items:center;bottom:0;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;left:0;position:fixed;right:0;top:0}.hc-modal{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;height:calc(100% - 50px);-webkit-box-pack:center;justify-content:center;max-height:calc(100% - 50px);min-height:135px}.hc-modal-auto{width:auto}.hc-modal-sm{width:300px}.hc-modal-md{width:500px}.hc-modal-lg{width:800px}.hc-modal-xl{width:992px}"]
                }] }
    ];
    /** @nocollapse */
    ModalWindowComponent.ctorParameters = function () { return [
        { type: ActiveModal },
        { type: ElementRef }
    ]; };
    ModalWindowComponent.propDecorators = {
        _ignoreOverlayClick: [{ type: Input }],
        _size: [{ type: Input }],
        _fadeInOut: [{ type: HostBinding, args: ['@fadeInOut',] }],
        _overlayClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
    };
    return ModalWindowComponent;
}());
export { ModalWindowComponent };
if (false) {
    /** @type {?} */
    ModalWindowComponent.prototype._ignoreOverlayClick;
    /** @type {?} */
    ModalWindowComponent.prototype._size;
    /**
     * @type {?}
     * @private
     */
    ModalWindowComponent.prototype.activeModal;
    /**
     * @type {?}
     * @private
     */
    ModalWindowComponent.prototype.el;
}
