// @healthcatalyst/cashmere/esm5/lib/modal/modal-overlay.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-overlay.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, HostListener, Input, ViewEncapsulation } from '@angular/core';
import { ActiveModal } from './active-modal';
import { animate, state, style, transition, trigger } from '@angular/animations';
var ModalOverlayComponent = /** @class */ (function () {
    function ModalOverlayComponent(activeModal) {
        this.activeModal = activeModal;
        this._ignoreEscapeKey = false;
    }
    /**
     * @return {?}
     */
    ModalOverlayComponent.prototype._fadeInOut = /**
     * @return {?}
     */
    function () {
        return state;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ModalOverlayComponent.prototype._escapeKey = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this._ignoreEscapeKey) {
            this.activeModal.dismiss();
        }
    };
    ModalOverlayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-modal-overlay',
                    template: '',
                    encapsulation: ViewEncapsulation.None,
                    // tslint:disable-next-line: no-host-metadata-property
                    host: { class: 'hc-modal-overlay' },
                    animations: [
                        trigger('fadeInOut', [
                            state('in', style({ opacity: 0.5 })),
                            transition('void <=> *', [
                                style({
                                    opacity: 0
                                }),
                                animate('0.2s ease-in-out')
                            ])
                        ])
                    ],
                    styles: ["\n            .hc-modal-overlay {\n                background-color: #000;\n                position: fixed;\n                top: 0;\n                left: 0;\n                right: 0;\n                bottom: 0;\n                opacity: 0.5;\n                display: block;\n            }\n        "]
                }] }
    ];
    /** @nocollapse */
    ModalOverlayComponent.ctorParameters = function () { return [
        { type: ActiveModal }
    ]; };
    ModalOverlayComponent.propDecorators = {
        _ignoreEscapeKey: [{ type: Input }],
        _fadeInOut: [{ type: HostBinding, args: ['@fadeInOut',] }],
        _escapeKey: [{ type: HostListener, args: ['document:keyup.escape', ['$event'],] }]
    };
    return ModalOverlayComponent;
}());
export { ModalOverlayComponent };
if (false) {
    /** @type {?} */
    ModalOverlayComponent.prototype._ignoreEscapeKey;
    /**
     * @type {?}
     * @private
     */
    ModalOverlayComponent.prototype.activeModal;
}
