// @healthcatalyst/cashmere/esm5/lib/modal/modal.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * @template T
 */
var /**
 * @template T
 */
HcModal = /** @class */ (function () {
    function HcModal() {
        this._result = new Subject();
        this._modalClose = new EventEmitter();
    }
    Object.defineProperty(HcModal.prototype, "result", {
        /** Subscribe to result in order to get access to modal result values passed in ActiveModal.close() */
        get: /**
         * Subscribe to result in order to get access to modal result values passed in ActiveModal.close()
         * @return {?}
         */
        function () {
            return this._result.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /** Closes the modal with a result.
     * Use this close method when opening a modal using a TemplateRef.
     * To close a modal that was created from a Component, inject ActiveModal and use the close method
     * on ActiveModal */
    /**
     * Closes the modal with a result.
     * Use this close method when opening a modal using a TemplateRef.
     * To close a modal that was created from a Component, inject ActiveModal and use the close method
     * on ActiveModal
     * @param {?=} result
     * @return {?}
     */
    HcModal.prototype.close = /**
     * Closes the modal with a result.
     * Use this close method when opening a modal using a TemplateRef.
     * To close a modal that was created from a Component, inject ActiveModal and use the close method
     * on ActiveModal
     * @param {?=} result
     * @return {?}
     */
    function (result) {
        this.removeModalElements();
        this._result.next(result);
    };
    /** Dismisses the modal with no result.
     * Use this dismiss method when opening a modal using a TemplateRef.
     * To dismiss a modal that was created from a Component, inject ActiveModal and use the dismiss method
     * on ActiveModal */
    /**
     * Dismisses the modal with no result.
     * Use this dismiss method when opening a modal using a TemplateRef.
     * To dismiss a modal that was created from a Component, inject ActiveModal and use the dismiss method
     * on ActiveModal
     * @return {?}
     */
    HcModal.prototype.dismiss = /**
     * Dismisses the modal with no result.
     * Use this dismiss method when opening a modal using a TemplateRef.
     * To dismiss a modal that was created from a Component, inject ActiveModal and use the dismiss method
     * on ActiveModal
     * @return {?}
     */
    function () {
        this.removeModalElements();
    };
    /**
     * @private
     * @return {?}
     */
    HcModal.prototype.removeModalElements = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.window) {
            /** @type {?} */
            var windowNativeElement = this.window.location.nativeElement;
            windowNativeElement.parentNode.removeChild(windowNativeElement);
            this.window.destroy();
        }
        if (this.overlay) {
            /** @type {?} */
            var overlayNativeElement = this.overlay.location.nativeElement;
            overlayNativeElement.parentNode.removeChild(overlayNativeElement);
            this.overlay.destroy();
        }
        if (this.componentRef) {
            /** @type {?} */
            var componentNativeElement = this.componentRef.location.nativeElement;
            componentNativeElement.parentNode.removeChild(componentNativeElement);
            this.componentRef.destroy();
        }
        if (this._removeOpenClass) {
            this._removeOpenClass();
        }
        this._modalClose.emit();
        this.window = null;
        this.overlay = null;
        this.componentRef = null;
        this._removeOpenClass = null;
    };
    return HcModal;
}());
/**
 * @template T
 */
export { HcModal };
if (false) {
    /**
     * Allows direct access to the component used to create the modal. Null when TemplateRef is used
     * @type {?}
     */
    HcModal.prototype.componentRef;
    /**
     * Allows direct access to overlay component which holds the component/template
     * @type {?}
     */
    HcModal.prototype.overlay;
    /**
     * Allows direct access to window component which holds the component/template
     * @type {?}
     */
    HcModal.prototype.window;
    /**
     * @type {?}
     * @private
     */
    HcModal.prototype._result;
    /** @type {?} */
    HcModal.prototype._removeOpenClass;
    /** @type {?} */
    HcModal.prototype._modalClose;
    /**
     * Data that was passed in through ModalOptions
     * @type {?}
     */
    HcModal.prototype.data;
}
