// @healthcatalyst/cashmere/esm2015/lib/navbar/navbar-link/navbar-link.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/navbar/navbar-link/navbar-link.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
/**
 * Primary navigation links
 */
export class NavbarLinkComponent {
    /**
     * @param {?} el
     * @param {?} ref
     */
    constructor(el, ref) {
        this.el = el;
        this.ref = ref;
        /**
         * Sets the RouterLinkActive options to match the url exactly to set active state. *Default is false.*
         * See https://angular.io/api/router/RouterLinkActive#description
         */
        this.exact = false;
        this._hidden = false;
    }
    /**
     * Disable visibility of component from view
     * @return {?}
     */
    hide() {
        this._hidden = true;
        this.ref.detectChanges();
    }
    /**
     * Enable visibility of component from view
     * @return {?}
     */
    show() {
        this._hidden = false;
        this.ref.detectChanges();
    }
    /**
     * @return {?}
     */
    _getWidth() {
        return this.el.nativeElement.scrollWidth;
    }
}
NavbarLinkComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-navbar-link',
                template: "<a *ngIf=\"!_hidden\" title=\"{{linkText}}\" [routerLink]=\"uri\" routerLinkActive=\"active\"\n   class=\"navbar-link\" [ngClass]=\"{ 'active': this.active === true, 'inactive': this.active === false }\"\n   [routerLinkActiveOptions]=\"{exact: exact}\">\n    {{linkText}}<ng-content></ng-content>\n</a>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
NavbarLinkComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
NavbarLinkComponent.propDecorators = {
    active: [{ type: Input }],
    uri: [{ type: Input }],
    linkText: [{ type: Input }],
    exact: [{ type: Input }]
};
if (false) {
    /**
     * (optional) forces active state *Default is `null`.*
     * @type {?}
     */
    NavbarLinkComponent.prototype.active;
    /**
     * RouterLink uri. See https://angular.io/api/router/RouterLink
     * @type {?}
     */
    NavbarLinkComponent.prototype.uri;
    /**
     * The text to display. This can also optionally be the contents within the element
     * @type {?}
     */
    NavbarLinkComponent.prototype.linkText;
    /**
     * Sets the RouterLinkActive options to match the url exactly to set active state. *Default is false.*
     * See https://angular.io/api/router/RouterLinkActive#description
     * @type {?}
     */
    NavbarLinkComponent.prototype.exact;
    /** @type {?} */
    NavbarLinkComponent.prototype._hidden;
    /**
     * @type {?}
     * @private
     */
    NavbarLinkComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    NavbarLinkComponent.prototype.ref;
}
