// @healthcatalyst/cashmere/esm5/lib/date-range/date-range/date-range.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-range/date-range/date-range.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Output, EventEmitter, Input, ElementRef, Directive, HostListener } from '@angular/core';
import { DatePipe } from '@angular/common';
import { CalendarOverlayService } from '../services/calendar-overlay.service';
import { ConfigStoreService } from '../services/config-store.service';
var DateRangeDirective = /** @class */ (function () {
    function DateRangeDirective(_elementRef, calendarOverlayService, configStoreService) {
        var _this = this;
        this._elementRef = _elementRef;
        this.calendarOverlayService = calendarOverlayService;
        this.configStoreService = configStoreService;
        /**
         * Emits when date range is been changed.
         */
        this.selectedDateRangeChanged = new EventEmitter();
        /**
         * Emits either a numerical index for the selected preset, or a `DateRange` if the selected value is not a preset
         */
        this.selectedPresetChanged = new EventEmitter();
        configStoreService.rangeUpdate$.subscribe((/**
         * @param {?} daterange
         * @return {?}
         */
        function (daterange) {
            _this.selectedDateRangeChanged.emit(daterange);
        }));
        configStoreService.presetUpdate$.subscribe((/**
         * @param {?} preset
         * @return {?}
         */
        function (preset) {
            _this.selectedPresetChanged.emit(preset);
        }));
    }
    /**
     * @return {?}
     */
    DateRangeDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    DateRangeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._overlayRef) {
            this._overlayRef.detach();
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    DateRangeDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes['options']) {
            /** @type {?} */
            var options = changes['options'].currentValue;
            this.configStoreService.updateDateRangeOptions(options);
        }
        if (changes['selectedDate']) {
            /** @type {?} */
            var selectedDate = changes['selectedDate'].currentValue;
            if (typeof selectedDate === 'number') {
                this.configStoreService.updatePreset(selectedDate);
            }
            else {
                this.configStoreService.updateRange(selectedDate);
            }
        }
    };
    /**
     * @return {?}
     */
    DateRangeDirective.prototype._onClick = /**
     * @return {?}
     */
    function () {
        this._overlayRef = this.calendarOverlayService.open(this._elementRef);
    };
    DateRangeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hcDateRange]',
                    providers: [CalendarOverlayService, ConfigStoreService, DatePipe]
                },] }
    ];
    /** @nocollapse */
    DateRangeDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: CalendarOverlayService },
        { type: ConfigStoreService }
    ]; };
    DateRangeDirective.propDecorators = {
        selectedDateRangeChanged: [{ type: Output }],
        selectedDate: [{ type: Input }],
        selectedPresetChanged: [{ type: Output }],
        options: [{ type: Input }],
        _onClick: [{ type: HostListener, args: ['click',] }]
    };
    return DateRangeDirective;
}());
export { DateRangeDirective };
if (false) {
    /**
     * Emits when date range is been changed.
     * @type {?}
     */
    DateRangeDirective.prototype.selectedDateRangeChanged;
    /**
     * Sets the selected date range. Accepts either a `DateRange` or a numerical index for preset.
     * @type {?}
     */
    DateRangeDirective.prototype.selectedDate;
    /**
     * Emits either a numerical index for the selected preset, or a `DateRange` if the selected value is not a preset
     * @type {?}
     */
    DateRangeDirective.prototype.selectedPresetChanged;
    /**
     * Configuration to setup behavior of component.
     * @type {?}
     */
    DateRangeDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype._overlayRef;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype.calendarOverlayService;
    /** @type {?} */
    DateRangeDirective.prototype.configStoreService;
}
