// @healthcatalyst/cashmere/esm2015/lib/app-switcher/app-switcher.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/app-switcher/app-switcher.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { APP_SWITCHER_CONFIG } from './app-switcher-interfaces';
/**
 * An app switcher service designed to work with the Health Catalyst DOS platform's DiscoveryService
 */
export class AppSwitcherService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        this.http = http;
        this.config = config;
        if (!config || !config.discoveryServiceUri) {
            throw new Error([
                'Failed to initialize AppSwitcherService: invalid APP_SWITCHER_CONFIG.',
                'You must provide a config object with a `discoveryServiceUri`.',
                `(value provided: ${config ? config.discoveryServiceUri : config})`
            ].join(' '));
        }
        this.discoveryServiceUri = this.normalizeUri(this.config.discoveryServiceUri);
        this.allApplicationsUri = `${this.discoveryServiceUri}/apps`;
    }
    /**
     * @return {?}
     */
    getApplications() {
        /** @type {?} */
        const url = `${this.discoveryServiceUri}/v1/Services?$filter=DiscoveryType eq 'Application' and IsHidden eq false&$top=12`;
        return this.http.get(url, { withCredentials: true });
    }
    /**
     * @private
     * @param {?} uri
     * @return {?}
     */
    normalizeUri(uri) {
        return uri.replace(/\/(v\d+\/?)?$/, '');
    }
}
AppSwitcherService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AppSwitcherService.ctorParameters = () => [
    { type: HttpClient },
    { type: undefined, decorators: [{ type: Inject, args: [APP_SWITCHER_CONFIG,] }] }
];
if (false) {
    /** @type {?} */
    AppSwitcherService.prototype.allApplicationsUri;
    /**
     * @type {?}
     * @private
     */
    AppSwitcherService.prototype.discoveryServiceUri;
    /**
     * @type {?}
     * @private
     */
    AppSwitcherService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AppSwitcherService.prototype.config;
}
