// @healthcatalyst/cashmere/esm2015/lib/app-switcher/app-switcher-interfaces.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/app-switcher/app-switcher-interfaces.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { of } from 'rxjs';
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function IDiscoveryApplication() { }
if (false) {
    /** @type {?} */
    IDiscoveryApplication.prototype.ServiceName;
    /** @type {?} */
    IDiscoveryApplication.prototype.FriendlyName;
    /** @type {?} */
    IDiscoveryApplication.prototype.Description;
    /** @type {?} */
    IDiscoveryApplication.prototype.BuildNumber;
    /** @type {?} */
    IDiscoveryApplication.prototype.Version;
    /** @type {?} */
    IDiscoveryApplication.prototype.DiscoveryServiceId;
    /** @type {?} */
    IDiscoveryApplication.prototype.ServiceUrl;
    /** @type {?|undefined} */
    IDiscoveryApplication.prototype.Heartbeat;
    /** @type {?|undefined} */
    IDiscoveryApplication.prototype.IsHidden;
    /** @type {?} */
    IDiscoveryApplication.prototype.DiscoveryType;
    /** @type {?} */
    IDiscoveryApplication.prototype.Icon;
}
/**
 * @record
 */
export function IDiscoveryRequest() { }
if (false) {
    /** @type {?} */
    IDiscoveryRequest.prototype.value;
}
/**
 * @record
 */
export function IAppSwitcherService() { }
if (false) {
    /** @type {?} */
    IAppSwitcherService.prototype.allApplicationsUri;
    /**
     * @return {?}
     */
    IAppSwitcherService.prototype.getApplications = function () { };
}
/** @type {?} */
export const APP_SWITCHER_SERVICE = new InjectionToken('IAppSwitcherService');
/**
 * @record
 */
export function IAppSwitcherConfig() { }
if (false) {
    /** @type {?} */
    IAppSwitcherConfig.prototype.discoveryServiceUri;
}
/** @type {?} */
export const APP_SWITCHER_CONFIG = new InjectionToken('IAppSwitcherConfig');
export class MockAppSwitcherService {
    /**
     * @return {?}
     */
    getApplications() {
        return of({
            value: [
                {
                    ServiceName: 'MyApp',
                    FriendlyName: 'My App',
                    Description: 'The currently showing app',
                    BuildNumber: '123',
                    Version: 1,
                    DiscoveryServiceId: 1,
                    ServiceUrl: 'http://myapp.com',
                    DiscoveryType: 'Application',
                    Icon: ''
                },
                {
                    ServiceName: 'AnotherApp',
                    FriendlyName: 'Another App',
                    Description: 'App we could switch to',
                    BuildNumber: '123',
                    Version: 1,
                    DiscoveryServiceId: 2,
                    ServiceUrl: 'http://anotherapp.com',
                    DiscoveryType: 'Application',
                    Icon: ''
                }
            ]
        });
    }
}
if (false) {
    /** @type {?} */
    MockAppSwitcherService.prototype.allApplicationsUri;
}
