// @healthcatalyst/cashmere/esm2015/lib/typeform-survey/typeform-survey.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/typeform-survey/typeform-survey.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * @return {?}
 */
export function throwErrorForMissingSurveyUri() {
    throw Error(`SurveyUri must be specified on element hc-typeform-survey`);
}
export class TypeformWindow extends Window {
}
if (false) {
    /** @type {?} */
    TypeformWindow.prototype.typeformEmbed;
}
export class TypeformSurveyComponent {
    constructor() {
        this._id = 'typef_orm_share';
    }
    /**
     * TypeForm survey URI you want to use. Example: https://somecompany.typeform.com/to/surveyId?parameter=parametervalue
     * @param {?} uri
     * @return {?}
     */
    set surveyUri(uri) {
        this._surveyUri = uri;
        this.refreshFullUri();
    }
    /**
     * @return {?}
     */
    get surveyUri() {
        return this._surveyUri;
    }
    /**
     * App version which will be passed to the survey in a hidden field. Ensures you know what version the feedback is referencing.
     * @param {?} version
     * @return {?}
     */
    set appVersion(version) {
        this._appVersion = version;
        this.refreshFullUri();
    }
    /**
     * @return {?}
     */
    get appVersion() {
        return this._appVersion;
    }
    /**
     * Opens the survey specified in the surveyUri
     * @return {?}
     */
    open() {
        if (!document.getElementById(this._id)) {
            this.getScripts();
        }
        else {
            ((/** @type {?} */ (window))).typeformEmbed.makePopup(this._fullUri, {
                mode: 'drawer_right',
                autoOpen: true,
                opacity: 100,
                autoClose: 0,
                hideScrollbars: true
            });
        }
    }
    /**
     * @private
     * @return {?}
     */
    refreshFullUri() {
        /** @type {?} */
        let varChar = this.surveyUri.includes('?') ? '&' : '?';
        this._fullUri = this.appVersion ? this.surveyUri + varChar + 'app_version=' + this.appVersion : this.surveyUri;
    }
    /**
     * @private
     * @return {?}
     */
    getScripts() {
        if (!this.surveyUri) {
            throwErrorForMissingSurveyUri();
        }
        /* this is directly from the embed
                   markup given from TypeForm */
        /** @type {?} */
        let embedScript;
        /** @type {?} */
        let firstScript;
        if (!document.getElementById(this._id)) {
            // create new embed script with typeform cdn source
            embedScript = document.createElement('script');
            embedScript.id = this._id;
            embedScript.src = `https://embed.typeform.com/embed.js`;
            // insert embed script before other js scripts
            firstScript = document.getElementsByTagName.call(document, 'script')[0];
            if (firstScript.parentNode) {
                firstScript.parentNode.insertBefore(embedScript, firstScript);
            }
        }
    }
}
TypeformSurveyComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-typeform-survey',
                template: `
        <a
            class="typeform-share link"
            [href]="_fullUri"
            data-mode="drawer_right"
            data-auto-open="true"
            data-submit-close-delay="0"
            target="_blank"
            rel="noopener"
        ></a>
    `,
                encapsulation: ViewEncapsulation.None
            }] }
];
TypeformSurveyComponent.propDecorators = {
    surveyUri: [{ type: Input }],
    appVersion: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TypeformSurveyComponent.prototype._fullUri;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._surveyUri;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._appVersion;
    /**
     * @type {?}
     * @private
     */
    TypeformSurveyComponent.prototype._id;
}
