// @healthcatalyst/cashmere/esm5/lib/date-range/services/calendar-overlay.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-range/services/calendar-overlay.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { Overlay, OverlayConfig, OverlayRef } from '@angular/cdk/overlay';
import { ComponentPortal, PortalInjector } from '@angular/cdk/portal';
import { PickerOverlayComponent } from '../picker-overlay/picker-overlay.component';
var CalendarOverlayService = /** @class */ (function () {
    function CalendarOverlayService(overlay, injector) {
        this.overlay = overlay;
        this.injector = injector;
    }
    /**
     * @param {?} hostElemRef
     * @return {?}
     */
    CalendarOverlayService.prototype.open = /**
     * @param {?} hostElemRef
     * @return {?}
     */
    function (hostElemRef) {
        this.hostElemRef = hostElemRef;
        /** @type {?} */
        var overlayRef = this._createOverlay();
        /** @type {?} */
        var portalInjector = this._createInjector(overlayRef);
        /** @type {?} */
        var calendarPortal = new ComponentPortal(PickerOverlayComponent, null, portalInjector);
        overlayRef.attach(calendarPortal);
        overlayRef.backdropClick().subscribe((/**
         * @return {?}
         */
        function () { return overlayRef.dispose(); }));
        return overlayRef;
    };
    /**
     * @private
     * @return {?}
     */
    CalendarOverlayService.prototype._createOverlay = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var overlayConfig = this._getOverlayConfig();
        return this.overlay.create(overlayConfig);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarOverlayService.prototype._getOverlayConfig = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var positionStrategy = this.overlay
            .position()
            .flexibleConnectedTo(this.hostElemRef)
            .withFlexibleDimensions(false)
            .withViewportMargin(8)
            .withDefaultOffsetY(12)
            .withLockedPosition()
            .withPositions([
            {
                originX: 'start',
                originY: 'bottom',
                overlayX: 'start',
                overlayY: 'top'
            },
            {
                originX: 'start',
                originY: 'top',
                overlayX: 'start',
                overlayY: 'bottom'
            },
            {
                originX: 'end',
                originY: 'bottom',
                overlayX: 'end',
                overlayY: 'top'
            },
            {
                originX: 'end',
                originY: 'top',
                overlayX: 'end',
                overlayY: 'bottom'
            }
        ]);
        /** @type {?} */
        var overlayConfig = new OverlayConfig({
            hasBackdrop: true,
            backdropClass: 'hc-overlay-transparent-backdrop',
            panelClass: 'hc-date-range-overlay',
            scrollStrategy: this.overlay.scrollStrategies.block(),
            positionStrategy: positionStrategy
        });
        return overlayConfig;
    };
    /**
     * @private
     * @param {?} overlayRef
     * @return {?}
     */
    CalendarOverlayService.prototype._createInjector = /**
     * @private
     * @param {?} overlayRef
     * @return {?}
     */
    function (overlayRef) {
        /** @type {?} */
        var injectionTokens = new WeakMap();
        injectionTokens.set(OverlayRef, overlayRef);
        return new PortalInjector(this.injector, injectionTokens);
    };
    CalendarOverlayService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CalendarOverlayService.ctorParameters = function () { return [
        { type: Overlay },
        { type: Injector }
    ]; };
    return CalendarOverlayService;
}());
export { CalendarOverlayService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CalendarOverlayService.prototype.hostElemRef;
    /**
     * @type {?}
     * @private
     */
    CalendarOverlayService.prototype.overlay;
    /**
     * @type {?}
     * @private
     */
    CalendarOverlayService.prototype.injector;
}
