// @healthcatalyst/cashmere/esm5/lib/date-range/services/config-store.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-range/services/config-store.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';
var ConfigStoreService = /** @class */ (function () {
    function ConfigStoreService() {
        var _this = this;
        this.defaultOptions = {
            presets: [],
            format: 'medium',
            mode: 'date',
            hourCycle: 12,
            excludeWeekends: false,
            locale: 'en-us',
            applyLabel: 'Apply',
            cancelLabel: 'Cancel',
            startDatePrefix: 'Start date:',
            endDatePrefix: 'End date:',
            invalidDateLabel: 'Please enter valid date'
        };
        this.dateRangeOptionsSubject = new BehaviorSubject(this.defaultOptions);
        this.rangeUpdateSubject = new BehaviorSubject({ fromDate: undefined, toDate: undefined });
        this.presetUpdateSubject = new BehaviorSubject({
            fromDate: undefined,
            toDate: undefined
        });
        this.weekendFilter = (/**
         * @return {?}
         */
        function () { return true; });
        this.emptyWeekendFilter = (/**
         * @return {?}
         */
        function () { return true; });
        this.excludeWeekendFilter = (/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            /** @type {?} */
            var day = d.getDay();
            return day !== 0 && day !== 6;
        });
        this.dateRangeOptions$ = this.dateRangeOptionsSubject.pipe(map((/**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            return tslib_1.__assign({}, _this.defaultOptions, options);
        })), tap((/**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            if (!!options.excludeWeekends) {
                _this.weekendFilter = _this.excludeWeekendFilter;
            }
            else {
                _this.weekendFilter = _this.emptyWeekendFilter;
            }
        })));
        this.rangeUpdate$ = this.rangeUpdateSubject.pipe();
        this.presetUpdate$ = this.presetUpdateSubject.pipe();
    }
    /**
     * @param {?} options
     * @return {?}
     */
    ConfigStoreService.prototype.updateDateRangeOptions = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.dateRangeOptionsSubject.next(options);
    };
    /**
     * @param {?} dateRange
     * @return {?}
     */
    ConfigStoreService.prototype.updateRange = /**
     * @param {?} dateRange
     * @return {?}
     */
    function (dateRange) {
        this.rangeUpdateSubject.next(dateRange);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ConfigStoreService.prototype.updatePreset = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.presetUpdateSubject.next(value);
    };
    ConfigStoreService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ConfigStoreService.ctorParameters = function () { return []; };
    return ConfigStoreService;
}());
export { ConfigStoreService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.defaultOptions;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.dateRangeOptionsSubject;
    /** @type {?} */
    ConfigStoreService.prototype.dateRangeOptions$;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.rangeUpdateSubject;
    /** @type {?} */
    ConfigStoreService.prototype.rangeUpdate$;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.presetUpdateSubject;
    /** @type {?} */
    ConfigStoreService.prototype.presetUpdate$;
    /** @type {?} */
    ConfigStoreService.prototype.weekendFilter;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.emptyWeekendFilter;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.excludeWeekendFilter;
}
