// @healthcatalyst/cashmere/esm2015/lib/stepper/stepper.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/stepper/stepper.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, ViewEncapsulation, Output, EventEmitter } from '@angular/core';
import { Router, NavigationEnd, UrlSerializer } from '@angular/router';
import { parseBooleanAttribute } from '../util';
/**
 * @param {?} stepsWithoutRouterLink
 * @return {?}
 */
export function throwErrorForMissingRouterLink(stepsWithoutRouterLink) {
    /** @type {?} */
    const stepLabels = stepsWithoutRouterLink.map((/**
     * @param {?} step
     * @return {?}
     */
    step => step.label));
    throw Error(`Routerlink missing on ${stepLabels.join(',')}`);
}
/**
 * @record
 */
export function StepInterface() { }
if (false) {
    /**
     * The text to be displayed under the step indicator
     * @type {?|undefined}
     */
    StepInterface.prototype.label;
    /**
     * If using a router, the route this step should navigate to
     * @type {?|undefined}
     */
    StepInterface.prototype.routerLink;
    /**
     * The icon set the step's hc-icon should use. Must be set if using the icon parameter.
     * @type {?|undefined}
     */
    StepInterface.prototype.iconSet;
    /**
     * An glyph that should be displayed inside the step's circle.
     * If not set, the step number will display in the circle instead.
     * @type {?|undefined}
     */
    StepInterface.prototype.icon;
    /**
     * If true, the step will not be clickable
     * @type {?|undefined}
     */
    StepInterface.prototype.disabled;
    /**
     * Apply query params to the routerLink
     * @type {?|undefined}
     */
    StepInterface.prototype.queryParams;
}
/**
 * Convey progress through numbered steps, providing a wizard-like workflow.
 */
export class StepperComponent {
    /**
     * @param {?} router
     * @param {?} urlSerializer
     */
    constructor(router, urlSerializer) {
        this.router = router;
        this.urlSerializer = urlSerializer;
        this._routerEnabled = false;
        /**
         * Sets the layout of the progress stepper. *Defaults to `arrow`.*
         */
        this.type = 'arrow';
        this._color = 'green';
        this._showStepCount = false;
        this._useRouterOutlet = true;
        this._activeIndex = undefined;
        /**
         * Emits the current zero-based index for the active step whenever it changes
         */
        this.activeIndexChange = new EventEmitter();
        this._hostClass = 'hc-stepper-' + this.color;
        this.router.events.forEach((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            if (event instanceof NavigationEnd) {
                this.findCurrentStep();
            }
        }));
    }
    /**
     * Sets the highlight color of the progress stepper. *Defaults to `green`.*
     * @return {?}
     */
    get color() {
        return this._color;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set color(value) {
        this._color = value;
        this._hostClass = 'hc-stepper-' + this._color;
    }
    /**
     * Determines whether numerals should be displayed on each step indicator for the isolated type.
     * *Defaults to `false`.*
     * @return {?}
     */
    get showStepCount() {
        return this._showStepCount;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set showStepCount(value) {
        this._showStepCount = parseBooleanAttribute(value);
    }
    /**
     * If true, include a router outlet with the component. *Defaults to `true`.*
     * @return {?}
     */
    get useRouterOutlet() {
        return this._useRouterOutlet;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set useRouterOutlet(value) {
        this._useRouterOutlet = parseBooleanAttribute(value);
    }
    /**
     * Get or set the currently selected zero-based index of the stepper
     * @return {?}
     */
    get activeIndex() {
        return this._activeIndex;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeIndex(value) {
        if (!this.steps) {
            return;
        }
        if (value !== undefined && (value < 0 || value >= this.steps.length)) {
            throw Error('The hc-stepper activeIndex value of ' + value + ' is out of bounds');
        }
        if (this._routerEnabled && value !== undefined) {
            this.router.navigate([this.steps[value].routerLink]);
        }
        else {
            this._activeIndex = value;
            this.activeIndexChange.emit(this._activeIndex);
        }
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._checkForRouterUse();
        if (this._routerEnabled) {
            this.findCurrentStep();
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    _stepClick(index) {
        if (!this._routerEnabled && index !== undefined && this.steps[index].disabled !== true) {
            this._activeIndex = index;
            this.activeIndexChange.emit(this._activeIndex);
        }
    }
    /**
     * @private
     * @return {?}
     */
    _checkForRouterUse() {
        /** @type {?} */
        const countUsingRouter = this.steps.filter((/**
         * @param {?} step
         * @return {?}
         */
        step => step.routerLink !== undefined)).length;
        if (countUsingRouter > 0 && countUsingRouter < this.steps.length) {
            /** @type {?} */
            const stepMissingRouterLink = this.steps.filter((/**
             * @param {?} step
             * @return {?}
             */
            step => step.routerLink === undefined));
            throwErrorForMissingRouterLink(stepMissingRouterLink);
        }
        else {
            this._routerEnabled = countUsingRouter === this.steps.length;
        }
    }
    /**
     * Can be used to activate the current step following \@Input() steps routerLink or queryParams updates
     * @return {?}
     */
    findCurrentStep() {
        /** @type {?} */
        const foundActiveRouteIndex = this.steps.findIndex((/**
         * @param {?} step
         * @return {?}
         */
        step => {
            /** @type {?} */
            const urlTree = this.router.createUrlTree([step.routerLink], { queryParams: step.queryParams });
            /** @type {?} */
            const stepURL = this.urlSerializer.serialize(urlTree);
            return this.router.isActive(stepURL, true);
        }));
        this._activeIndex = foundActiveRouteIndex > -1 ? foundActiveRouteIndex : undefined;
        this.activeIndexChange.emit(this._activeIndex);
    }
}
StepperComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-stepper',
                template: "<div [ngClass]=\"'hc-stepper-' + type\">\n    <div class=\"hc-stepper-track\">\n        <ng-content select=\".before-steps-content\"></ng-content>\n        <div\n            class=\"hc-step\"\n            *ngFor=\"let step of steps; index as i; last as isLast; first as isFirst\"\n            [routerLink]=\"step.routerLink\"\n            [routerLinkActiveOptions]=\"{exact: true}\"\n            routerLinkActive=\"active\"\n            [queryParams]=\"step.queryParams\"\n            [ngClass]=\"{\n                'hc-step-arrow-active': type === 'arrow' && i === activeIndex,\n                'hc-step-arrow-incomplete': type === 'arrow' && (activeIndex === undefined || i > activeIndex),\n                'hc-step-arrow-complete': type === 'arrow' && (activeIndex !== undefined && i < activeIndex - 1),\n                'hc-step-arrow-completed': type === 'arrow' && (activeIndex !== undefined && i === activeIndex - 1),\n                'hc-step-isolated-active': type === 'isolated' && i === activeIndex,\n                'hc-step-isolated-incomplete': type === 'isolated' && (activeIndex === undefined || i > activeIndex),\n                'hc-step-isolated-complete': type === 'isolated' && (activeIndex !== undefined && i < activeIndex - 1),\n                'hc-step-isolated-completed': type === 'isolated' && (activeIndex !== undefined && i === activeIndex - 1),\n                'hc-step-last': isLast,\n                'hc-step-first': isFirst,\n                'hc-stepper-clickable': step.disabled !== undefined ? !step.disabled : true,\n                'hc-stepper-disabled': step.disabled !== undefined ? step.disabled : false\n            }\"\n            (click)=\"_stepClick(i)\"\n        >\n            <span class=\"hc-step-disk\">\n                <ng-container *ngIf=\"step.iconSet && step.icon; else elseBlock\">\n                    <hc-icon [fontSet]=\"step.iconSet\" [fontIcon]=\"step.icon\" class=\"hc-step-icon\"></hc-icon>\n                </ng-container>\n                <ng-template #elseBlock>{{ i + 1 }}</ng-template>\n            </span>\n            <span class=\"hc-step-label-count\" *ngIf=\"showStepCount\">{{ i + 1 }}</span>\n            <span class=\"hc-step-label\">{{ step.label }}</span>\n            <span class=\"hc-step-arrow\"></span>\n        </div>\n        <ng-content select=\".after-steps-content\"></ng-content>\n    </div>\n    <div *ngIf=\"useRouterOutlet\" class=\"hc-step-content\">\n        <router-outlet *ngIf=\"_routerEnabled\"></router-outlet>\n    </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".hc-step{-webkit-box-flex:1;flex:1;position:relative}.hc-stepper-arrow .hc-step{line-height:4rem;padding-left:3.6rem;white-space:nowrap;border-top-color:#f0f3f6}.hc-stepper-isolated .hc-step{padding-top:2rem;text-align:center}.hc-stepper-isolated .hc-step:before{background:#d7dde4;content:'';display:block;height:.5rem;left:0;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1}.hc-stepper-arrow .hc-step-disk{border-radius:50%;display:block;font-size:.9rem;height:1.6rem;left:1.2rem;line-height:1.6rem;position:absolute;text-align:center;top:1.2rem;width:1.6rem;box-shadow:0 0 0 3px #d7dde4;background-color:#fff}.hc-stepper-isolated .hc-step-disk{border-radius:50%;display:block;font-size:1rem;height:2rem;left:50%;line-height:2rem;margin-left:-1rem;position:absolute;text-align:center;top:0;width:2rem;z-index:2;box-shadow:0 0 0 3px #f0f3f6;background-color:#fff}.hc-stepper-green .hc-step-arrow-active .hc-step-disk,.hc-stepper-green .hc-step-isolated-active .hc-step-disk,.hc-stepper-green .hc-step-isolated-complete .hc-step-disk,.hc-stepper-green .hc-step-isolated-completed .hc-step-disk{box-shadow:0 0 0 3px #00a859}.hc-stepper-blue .hc-step-arrow-active .hc-step-disk,.hc-stepper-blue .hc-step-isolated-active .hc-step-disk,.hc-stepper-blue .hc-step-isolated-complete .hc-step-disk,.hc-stepper-blue .hc-step-isolated-completed .hc-step-disk{box-shadow:0 0 0 3px #00aeff}.hc-stepper-purple .hc-step-arrow-active .hc-step-disk,.hc-stepper-purple .hc-step-isolated-active .hc-step-disk,.hc-stepper-purple .hc-step-isolated-complete .hc-step-disk,.hc-stepper-purple .hc-step-isolated-completed .hc-step-disk{box-shadow:0 0 0 3px #6e53a3}.hc-stepper-orange .hc-step-arrow-active .hc-step-disk,.hc-stepper-orange .hc-step-isolated-active .hc-step-disk,.hc-stepper-orange .hc-step-isolated-complete .hc-step-disk,.hc-stepper-orange .hc-step-isolated-completed .hc-step-disk{box-shadow:0 0 0 3px #f8961d}.hc-stepper-red .hc-step-arrow-active .hc-step-disk,.hc-stepper-red .hc-step-isolated-active .hc-step-disk,.hc-stepper-red .hc-step-isolated-complete .hc-step-disk,.hc-stepper-red .hc-step-isolated-completed .hc-step-disk{box-shadow:0 0 0 3px #f13c45}.hc-stepper-green .hc-step-arrow-complete .hc-step-disk,.hc-stepper-green .hc-step-arrow-completed .hc-step-disk{box-shadow:0 0 0 3px #23c077}.hc-stepper-blue .hc-step-arrow-complete .hc-step-disk,.hc-stepper-blue .hc-step-arrow-completed .hc-step-disk{box-shadow:0 0 0 3px #42c3ff}.hc-stepper-purple .hc-step-arrow-complete .hc-step-disk,.hc-stepper-purple .hc-step-arrow-completed .hc-step-disk{box-shadow:0 0 0 3px #8973b4}.hc-stepper-orange .hc-step-arrow-complete .hc-step-disk,.hc-stepper-orange .hc-step-arrow-completed .hc-step-disk{box-shadow:0 0 0 3px #f9ab49}.hc-stepper-red .hc-step-arrow-complete .hc-step-disk,.hc-stepper-red .hc-step-arrow-completed .hc-step-disk{box-shadow:0 0 0 3px #f57178}.hc-step-arrow-active:not(.hc-step-last) .hc-step-arrow,.hc-step-arrow-incomplete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #f0f3f6;border-bottom:2rem solid #f0f3f6;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-green .hc-step-arrow-completed .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #00a859;border-bottom:2rem solid #00a859;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-blue .hc-step-arrow-completed .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #00aeff;border-bottom:2rem solid #00aeff;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-purple .hc-step-arrow-completed .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #6e53a3;border-bottom:2rem solid #6e53a3;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-orange .hc-step-arrow-completed .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #f8961d;border-bottom:2rem solid #f8961d;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-red .hc-step-arrow-completed .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #f13c45;border-bottom:2rem solid #f13c45;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-green .hc-step-arrow-complete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #23c077;border-bottom:2rem solid #23c077;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-blue .hc-step-arrow-complete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #42c3ff;border-bottom:2rem solid #42c3ff;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-purple .hc-step-arrow-complete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #8973b4;border-bottom:2rem solid #8973b4;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-orange .hc-step-arrow-complete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #f9ab49;border-bottom:2rem solid #f9ab49;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-red .hc-step-arrow-complete:not(.hc-step-last) .hc-step-arrow{border-left:1rem solid transparent;border-top:2rem solid #f57178;border-bottom:2rem solid #f57178;display:block;height:0;position:absolute;right:0;top:0;z-index:2}.hc-stepper-arrow .hc-step-label-count{display:none}.hc-stepper-isolated .hc-step-label-count{color:#333;display:inline-block;line-height:2rem;padding:0 .15rem}.hc-stepper-isolated .hc-step-label-count:after{content:'.'}.hc-step-isolated-active .hc-step-label-count{font-weight:600}.hc-stepper-arrow .hc-step-label{color:#fff}.hc-step-arrow-incomplete .hc-step-label{color:unset}.hc-stepper-isolated .hc-step-label{color:#333;display:inline-block;line-height:2rem;padding:0 .15rem}.hc-step-isolated-active .hc-step-label{font-weight:600}.hc-stepper-arrow .hc-step-icon{font-size:1rem;height:100%;line-height:1.6rem;width:100%}.hc-stepper-isolated .hc-step-icon{font-size:1.4rem;height:100%;line-height:1.92rem;width:100%}.hc-step-arrow-incomplete .hc-step-icon,.hc-step-isolated-incomplete .hc-step-icon{color:#d7dde4}.hc-stepper-green .hc-step-arrow-active .hc-step-icon,.hc-stepper-green .hc-step-isolated-active .hc-step-icon,.hc-stepper-green .hc-step-isolated-complete .hc-step-icon,.hc-stepper-green .hc-step-isolated-completed .hc-step-icon{color:#00a859}.hc-stepper-blue .hc-step-arrow-active .hc-step-icon,.hc-stepper-blue .hc-step-isolated-active .hc-step-icon,.hc-stepper-blue .hc-step-isolated-complete .hc-step-icon,.hc-stepper-blue .hc-step-isolated-completed .hc-step-icon{color:#00aeff}.hc-stepper-purple .hc-step-arrow-active .hc-step-icon,.hc-stepper-purple .hc-step-isolated-active .hc-step-icon,.hc-stepper-purple .hc-step-isolated-complete .hc-step-icon,.hc-stepper-purple .hc-step-isolated-completed .hc-step-icon{color:#6e53a3}.hc-stepper-orange .hc-step-arrow-active .hc-step-icon,.hc-stepper-orange .hc-step-isolated-active .hc-step-icon,.hc-stepper-orange .hc-step-isolated-complete .hc-step-icon,.hc-stepper-orange .hc-step-isolated-completed .hc-step-icon{color:#f8961d}.hc-stepper-red .hc-step-arrow-active .hc-step-icon,.hc-stepper-red .hc-step-isolated-active .hc-step-icon,.hc-stepper-red .hc-step-isolated-complete .hc-step-icon,.hc-stepper-red .hc-step-isolated-completed .hc-step-icon{color:#f13c45}.hc-stepper-green .hc-step-arrow-complete .hc-step-icon,.hc-stepper-green .hc-step-arrow-completed .hc-step-icon{color:#23c077}.hc-stepper-blue .hc-step-arrow-complete .hc-step-icon,.hc-stepper-blue .hc-step-arrow-completed .hc-step-icon{color:#42c3ff}.hc-stepper-purple .hc-step-arrow-complete .hc-step-icon,.hc-stepper-purple .hc-step-arrow-completed .hc-step-icon{color:#8973b4}.hc-stepper-orange .hc-step-arrow-complete .hc-step-icon,.hc-stepper-orange .hc-step-arrow-completed .hc-step-icon{color:#f9ab49}.hc-stepper-red .hc-step-arrow-complete .hc-step-icon,.hc-stepper-red .hc-step-arrow-completed .hc-step-icon{color:#f57178}.hc-stepper-clickable{cursor:pointer;outline:0}.hc-stepper-disabled{pointer-events:none}.hc-stepper-track{display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row}.hc-stepper-arrow .hc-stepper-track{background-color:#f0f3f6}.hc-stepper-isolated .hc-step-first:before{left:50%}.hc-stepper-green .hc-step-arrow-active{background-color:#00a859}.hc-stepper-blue .hc-step-arrow-active{background-color:#00aeff}.hc-stepper-purple .hc-step-arrow-active{background-color:#6e53a3}.hc-stepper-orange .hc-step-arrow-active{background-color:#f8961d}.hc-stepper-red .hc-step-arrow-active{background-color:#f13c45}.hc-stepper-green .hc-step-isolated-active:before{background:#00a859;right:50%}.hc-stepper-green .hc-step-isolated-active:after{background:#d7dde4;content:'';display:block;height:.5rem;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1;left:50%}.hc-stepper-blue .hc-step-isolated-active:before{background:#00aeff;right:50%}.hc-stepper-blue .hc-step-isolated-active:after{background:#d7dde4;content:'';display:block;height:.5rem;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1;left:50%}.hc-stepper-purple .hc-step-isolated-active:before{background:#6e53a3;right:50%}.hc-stepper-purple .hc-step-isolated-active:after{background:#d7dde4;content:'';display:block;height:.5rem;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1;left:50%}.hc-stepper-orange .hc-step-isolated-active:before{background:#f8961d;right:50%}.hc-stepper-orange .hc-step-isolated-active:after{background:#d7dde4;content:'';display:block;height:.5rem;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1;left:50%}.hc-stepper-red .hc-step-isolated-active:before{background:#f13c45;right:50%}.hc-stepper-red .hc-step-isolated-active:after{background:#d7dde4;content:'';display:block;height:.5rem;margin-top:-.25rem;position:absolute;right:0;top:1rem;z-index:1;left:50%}.hc-stepper-isolated .hc-step-last:before{right:50%}.hc-stepper-isolated .hc-step-last:after{display:none}.hc-stepper-green .hc-step-arrow-complete,.hc-stepper-green .hc-step-arrow-completed{background-color:#23c077}.hc-stepper-blue .hc-step-arrow-complete,.hc-stepper-blue .hc-step-arrow-completed{background-color:#42c3ff}.hc-stepper-purple .hc-step-arrow-complete,.hc-stepper-purple .hc-step-arrow-completed{background-color:#8973b4}.hc-stepper-orange .hc-step-arrow-complete,.hc-stepper-orange .hc-step-arrow-completed{background-color:#f9ab49}.hc-stepper-red .hc-step-arrow-complete,.hc-stepper-red .hc-step-arrow-completed{background-color:#f57178}.hc-step-arrow-complete:not(.hc-step-last):after,.hc-step-arrow-incomplete:not(.hc-step-last):after{border-right:.2rem solid #d7dde4;border-left:1rem solid transparent;border-top:2rem solid #d7dde4;border-bottom:2rem solid #d7dde4;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}.hc-stepper-green .hc-step-isolated-complete:before,.hc-stepper-green .hc-step-isolated-completed:before{background:#00a859}.hc-stepper-blue .hc-step-isolated-complete:before,.hc-stepper-blue .hc-step-isolated-completed:before{background:#00aeff}.hc-stepper-purple .hc-step-isolated-complete:before,.hc-stepper-purple .hc-step-isolated-completed:before{background:#6e53a3}.hc-stepper-orange .hc-step-isolated-complete:before,.hc-stepper-orange .hc-step-isolated-completed:before{background:#f8961d}.hc-stepper-red .hc-step-isolated-complete:before,.hc-stepper-red .hc-step-isolated-completed:before{background:#f13c45}.hc-stepper-green .hc-step-arrow-complete:not(.hc-step-last):not(.hc-step-arrow-completed):after{border-right:.2rem solid #00a859;border-left:1rem solid transparent;border-top:2rem solid #00a859;border-bottom:2rem solid #00a859;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}.hc-stepper-blue .hc-step-arrow-complete:not(.hc-step-last):not(.hc-step-arrow-completed):after{border-right:.2rem solid #00aeff;border-left:1rem solid transparent;border-top:2rem solid #00aeff;border-bottom:2rem solid #00aeff;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}.hc-stepper-purple .hc-step-arrow-complete:not(.hc-step-last):not(.hc-step-arrow-completed):after{border-right:.2rem solid #6e53a3;border-left:1rem solid transparent;border-top:2rem solid #6e53a3;border-bottom:2rem solid #6e53a3;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}.hc-stepper-orange .hc-step-arrow-complete:not(.hc-step-last):not(.hc-step-arrow-completed):after{border-right:.2rem solid #f8961d;border-left:1rem solid transparent;border-top:2rem solid #f8961d;border-bottom:2rem solid #f8961d;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}.hc-stepper-red .hc-step-arrow-complete:not(.hc-step-last):not(.hc-step-arrow-completed):after{border-right:.2rem solid #f13c45;border-left:1rem solid transparent;border-top:2rem solid #f13c45;border-bottom:2rem solid #f13c45;display:block;height:0;position:absolute;right:0;top:0;content:'';z-index:1}"]
            }] }
];
/** @nocollapse */
StepperComponent.ctorParameters = () => [
    { type: Router },
    { type: UrlSerializer }
];
StepperComponent.propDecorators = {
    steps: [{ type: Input }],
    type: [{ type: Input }],
    color: [{ type: Input }],
    showStepCount: [{ type: Input }],
    useRouterOutlet: [{ type: Input }],
    activeIndex: [{ type: Input }],
    activeIndexChange: [{ type: Output }],
    _hostClass: [{ type: HostBinding, args: ['class',] }]
};
if (false) {
    /** @type {?} */
    StepperComponent.prototype._routerEnabled;
    /**
     * An array defining the steps in the stepper
     * @type {?}
     */
    StepperComponent.prototype.steps;
    /**
     * Sets the layout of the progress stepper. *Defaults to `arrow`.*
     * @type {?}
     */
    StepperComponent.prototype.type;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype._color;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype._showStepCount;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype._useRouterOutlet;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype._activeIndex;
    /**
     * Emits the current zero-based index for the active step whenever it changes
     * @type {?}
     */
    StepperComponent.prototype.activeIndexChange;
    /** @type {?} */
    StepperComponent.prototype._hostClass;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    StepperComponent.prototype.urlSerializer;
}
