// @healthcatalyst/cashmere/esm5/lib/table/row.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/table/row.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// tslint:disable:directive-class-suffix use-input-property-decorator component-class-suffix
// tslint:disable:no-host-metadata-property no-inputs-metadata-property
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, Directive, ViewEncapsulation } from '@angular/core';
import { CDK_ROW_TEMPLATE, CdkFooterRow, CdkFooterRowDef, CdkHeaderRow, CdkHeaderRowDef, CdkRow, CdkRowDef } from '@angular/cdk/table';
/**
 * Header row definition for the hc-table.
 * Captures the header row's template and other header properties such as the columns to display.
 */
var HcHeaderRowDef = /** @class */ (function (_super) {
    tslib_1.__extends(HcHeaderRowDef, _super);
    function HcHeaderRowDef() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcHeaderRowDef.decorators = [
        { type: Directive, args: [{
                    selector: '[hcHeaderRowDef]',
                    providers: [{ provide: CdkHeaderRowDef, useExisting: HcHeaderRowDef }],
                    inputs: ['columns: hcHeaderRowDef', 'sticky: hcHeaderRowDefSticky']
                },] }
    ];
    return HcHeaderRowDef;
}(CdkHeaderRowDef));
export { HcHeaderRowDef };
/**
 * Footer row definition for the mat-table.
 * Captures the footer row's template and other footer properties such as the columns to display.
 */
var HcFooterRowDef = /** @class */ (function (_super) {
    tslib_1.__extends(HcFooterRowDef, _super);
    function HcFooterRowDef() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcFooterRowDef.decorators = [
        { type: Directive, args: [{
                    selector: '[hcFooterRowDef]',
                    providers: [{ provide: CdkFooterRowDef, useExisting: HcFooterRowDef }],
                    inputs: ['columns: hcFooterRowDef', 'sticky: hcFooterRowDefSticky']
                },] }
    ];
    return HcFooterRowDef;
}(CdkFooterRowDef));
export { HcFooterRowDef };
/**
 * Data row definition for the mat-table.
 * Captures the footer row's template and other footer properties such as the columns to display and
 * a when predicate that describes when this row should be used.
 * @template T
 */
var HcRowDef = /** @class */ (function (_super) {
    tslib_1.__extends(HcRowDef, _super);
    function HcRowDef() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcRowDef.decorators = [
        { type: Directive, args: [{
                    selector: '[hcRowDef]',
                    providers: [{ provide: CdkRowDef, useExisting: HcRowDef }],
                    inputs: ['columns: hcRowDefColumns', 'when: hcRowDefWhen']
                },] }
    ];
    return HcRowDef;
}(CdkRowDef));
export { HcRowDef };
/**
 * Footer template container that contains the cell outlet. Adds the right class and role.
 */
var HcHeaderRow = /** @class */ (function (_super) {
    tslib_1.__extends(HcHeaderRow, _super);
    function HcHeaderRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcHeaderRow.decorators = [
        { type: Component, args: [{
                    selector: 'hc-header-row, tr[hc-header-row]',
                    template: CDK_ROW_TEMPLATE,
                    host: {
                        class: 'hc-header-row',
                        role: 'row'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'hcHeaderRow',
                    providers: [{ provide: CdkHeaderRow, useExisting: HcHeaderRow }]
                }] }
    ];
    return HcHeaderRow;
}(CdkHeaderRow));
export { HcHeaderRow };
/**
 * Footer template container that contains the cell outlet. Adds the right class and role.
 */
var HcFooterRow = /** @class */ (function (_super) {
    tslib_1.__extends(HcFooterRow, _super);
    function HcFooterRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcFooterRow.decorators = [
        { type: Component, args: [{
                    selector: 'hc-footer-row, tr[hc-footer-row]',
                    template: CDK_ROW_TEMPLATE,
                    host: {
                        class: 'hc-footer-row',
                        role: 'row'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'hcFooterRow',
                    providers: [{ provide: CdkFooterRow, useExisting: HcFooterRow }]
                }] }
    ];
    return HcFooterRow;
}(CdkFooterRow));
export { HcFooterRow };
/**
 * Data row template container that contains the cell outlet. Adds the right class and role.
 */
var HcRow = /** @class */ (function (_super) {
    tslib_1.__extends(HcRow, _super);
    function HcRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HcRow.decorators = [
        { type: Component, args: [{
                    selector: 'hc-row, tr[hc-row]',
                    template: CDK_ROW_TEMPLATE,
                    host: {
                        class: 'hc-row',
                        role: 'row'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'hcRow',
                    providers: [{ provide: CdkRow, useExisting: HcRow }]
                }] }
    ];
    return HcRow;
}(CdkRow));
export { HcRow };
