// @healthcatalyst/cashmere/esm5/lib/picklist/services/picklist-state.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/picklist/services/picklist-state.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { PicklistOptionsSource, PicklistSettings } from '../picklist.model';
import { FilterableSelectList, ValueListOption, ValueSetListOption } from '../pane/picklist-pane.model';
import { PicklistFilterLocalService } from './picklist-filter-local.service';
var PicklistStateService = /** @class */ (function () {
    function PicklistStateService(localFilterService) {
        this.localFilterService = localFilterService;
        this.picklist = new PicklistSettings();
        this.optionsSource = new PicklistOptionsSource();
        this.valueList = new FilterableSelectList();
        this.valueSetList = new FilterableSelectList();
    }
    /**
     * @param {?} settings
     * @param {?} optionsSource
     * @param {?} pane
     * @return {?}
     */
    PicklistStateService.prototype.reset = /**
     * @param {?} settings
     * @param {?} optionsSource
     * @param {?} pane
     * @return {?}
     */
    function (settings, optionsSource, pane) {
        this.picklist = settings;
        this.pane = pane;
        this.optionsSource = optionsSource;
        this.clearList(this.valueList);
        this.clearList(this.valueSetList);
    };
    /**
     * @param {?} options
     * @param {?=} searchTokens
     * @return {?}
     */
    PicklistStateService.prototype.updateValueList = /**
     * @param {?} options
     * @param {?=} searchTokens
     * @return {?}
     */
    function (options, searchTokens) {
        if (searchTokens === void 0) { searchTokens = []; }
        /** @type {?} */
        var listOptions = options.map((/**
         * @param {?} v
         * @return {?}
         */
        function (v) { return new ValueListOption(v, v.code); }));
        /** @type {?} */
        var companionList = this.pane.companion ? this.pane.companion.valueList : null;
        this.valueList.sort = this.picklist.sort;
        this.updateList(listOptions, this.valueList, companionList, searchTokens);
    };
    /**
     * @param {?} options
     * @param {?=} searchTokens
     * @return {?}
     */
    PicklistStateService.prototype.updateValueSetList = /**
     * @param {?} options
     * @param {?=} searchTokens
     * @return {?}
     */
    function (options, searchTokens) {
        var _this = this;
        if (searchTokens === void 0) { searchTokens = []; }
        /** @type {?} */
        var listOptions = new Array();
        /** @type {?} */
        var companionList = this.pane.companion ? this.pane.companion.valueSetList : null;
        options.forEach((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            /** @type {?} */
            var listOption = new ValueSetListOption(v, v.code);
            if (v.subValues && v.subValues.length > 0) {
                /** @type {?} */
                var subValueListOptions = v.subValues.map((/**
                 * @param {?} sv
                 * @return {?}
                 */
                function (sv) { return new ValueListOption(sv, sv.code); }));
                _this.updateList(subValueListOptions, listOption.subValuesSelectList, null, searchTokens);
            }
            listOptions.push(listOption);
        }));
        this.updateList(listOptions, this.valueSetList, companionList, searchTokens);
    };
    /**
     * @template T
     * @param {?} list
     * @return {?}
     */
    PicklistStateService.prototype.clearList = /**
     * @template T
     * @param {?} list
     * @return {?}
     */
    function (list) {
        list.options.clear();
        list.filteredOptions.length = 0;
        list.selectedOptions.clear();
        list.lastClickedOption = null;
        list.additionalRemoteOptions = 0;
    };
    /**
     * @private
     * @template T
     * @param {?} options
     * @param {?} list
     * @param {?=} companionList
     * @param {?=} searchTokens
     * @return {?}
     */
    PicklistStateService.prototype.updateList = /**
     * @private
     * @template T
     * @param {?} options
     * @param {?} list
     * @param {?=} companionList
     * @param {?=} searchTokens
     * @return {?}
     */
    function (options, list, companionList, searchTokens) {
        if (companionList === void 0) { companionList = null; }
        if (searchTokens === void 0) { searchTokens = []; }
        if (!this.optionsSource.isPaged && this.pane.shouldExcludeCompanion && companionList) {
            options = options.filter((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return !companionList.options.get(o.code); }));
        }
        options.forEach((/**
         * @param {?} o
         * @return {?}
         */
        function (o) {
            list.options.set(o.code, o);
            list.filteredOptions.push(o);
        }));
        this.localFilterService.filter(list, searchTokens);
    };
    PicklistStateService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PicklistStateService.ctorParameters = function () { return [
        { type: PicklistFilterLocalService }
    ]; };
    return PicklistStateService;
}());
export { PicklistStateService };
if (false) {
    /** @type {?} */
    PicklistStateService.prototype.pane;
    /** @type {?} */
    PicklistStateService.prototype.picklist;
    /** @type {?} */
    PicklistStateService.prototype.optionsSource;
    /** @type {?} */
    PicklistStateService.prototype.valueList;
    /** @type {?} */
    PicklistStateService.prototype.valueSetList;
    /**
     * @type {?}
     * @private
     */
    PicklistStateService.prototype.localFilterService;
}
