// @healthcatalyst/cashmere/esm2015/lib/modal/modal-window.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-window.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { animate, state, style, transition, trigger } from '@angular/animations';
import { Component, ElementRef, HostBinding, HostListener, Input, ViewEncapsulation } from '@angular/core';
import { ActiveModal } from './active-modal';
export class ModalWindowComponent {
    /**
     * @param {?} activeModal
     * @param {?} el
     */
    constructor(activeModal, el) {
        this.activeModal = activeModal;
        this.el = el;
        this._ignoreOverlayClick = false;
        this._size = 'auto';
    }
    /**
     * @return {?}
     */
    _fadeInOut() {
        return state;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _overlayClick(event) {
        /** @type {?} */
        let modalContentNotPresent = true;
        /** @type {?} */
        let path = this._eventPath(event);
        /** @type {?} */
        let modalWindowTargetIncluded = path.findIndex((/**
         * @param {?} p
         * @return {?}
         */
        p => p === this.el.nativeElement)) > -1;
        /** @type {?} */
        let classList = path.map((/**
         * @param {?} p
         * @return {?}
         */
        p => p.classList));
        for (let cl of classList) {
            if (cl) {
                if (cl.contains('hc-modal-content')) {
                    modalContentNotPresent = false;
                }
            }
        }
        /* The hc-modal has 100% height (although not completely visible).
            To enable closing the modal while clicking above or below the modal (in
            addition to the sides) this function will check to see if the click
            event includes:
                1. This window element
                2. Not the hc-modal-content and
                3. the overlay click option is disabled. */
        if (!this._ignoreOverlayClick && modalContentNotPresent && modalWindowTargetIncluded) {
            this.activeModal.dismiss();
        }
    }
    // Serves as a polyfill for Event.composedPath() or Event.Path
    /**
     * @param {?} evt
     * @return {?}
     */
    _eventPath(evt) {
        /** @type {?} */
        let path = (evt.composedPath && evt.composedPath()) || evt.path;
        /** @type {?} */
        let target = evt.target;
        if (path != null) {
            // Safari doesn't include Window, but it should.
            return path.indexOf(window) < 0 ? path.concat(window) : path;
        }
        if (target === window) {
            return [window];
        }
        /**
         * @param {?} node
         * @param {?=} memo
         * @return {?}
         */
        function _getParents(node, memo) {
            memo = memo || [];
            /** @type {?} */
            let parentNode = node.parentNode;
            if (!parentNode) {
                return memo;
            }
            else {
                return _getParents(parentNode, memo.concat(parentNode));
            }
        }
        return [target].concat(_getParents(target), window);
    }
}
ModalWindowComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-modal-window',
                template: `
        <div [class]="'hc-modal hc-modal-' + _size"><ng-content></ng-content></div>
    `,
                encapsulation: ViewEncapsulation.None,
                // tslint:disable-next-line: no-host-metadata-property
                host: { class: 'hc-modal-window' },
                animations: [
                    trigger('fadeInOut', [
                        state('in', style({ opacity: 1 })),
                        transition('void <=> *', [
                            style({
                                opacity: 0
                            }),
                            animate('0.2s ease-in-out')
                        ])
                    ])
                ],
                styles: [".hc-modal-window{-webkit-box-align:center;align-items:center;bottom:0;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;left:0;position:fixed;right:0;top:0}.hc-modal{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;height:calc(100% - 50px);-webkit-box-pack:center;justify-content:center;max-height:calc(100% - 50px);min-height:135px}.hc-modal-auto{width:auto}.hc-modal-sm{width:300px}.hc-modal-md{width:500px}.hc-modal-lg{width:800px}.hc-modal-xl{width:992px}"]
            }] }
];
/** @nocollapse */
ModalWindowComponent.ctorParameters = () => [
    { type: ActiveModal },
    { type: ElementRef }
];
ModalWindowComponent.propDecorators = {
    _ignoreOverlayClick: [{ type: Input }],
    _size: [{ type: Input }],
    _fadeInOut: [{ type: HostBinding, args: ['@fadeInOut',] }],
    _overlayClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    ModalWindowComponent.prototype._ignoreOverlayClick;
    /** @type {?} */
    ModalWindowComponent.prototype._size;
    /**
     * @type {?}
     * @private
     */
    ModalWindowComponent.prototype.activeModal;
    /**
     * @type {?}
     * @private
     */
    ModalWindowComponent.prototype.el;
}
