// @healthcatalyst/cashmere/esm2015/lib/modal/modal-overlay.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/modal/modal-overlay.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, HostListener, Input, ViewEncapsulation } from '@angular/core';
import { ActiveModal } from './active-modal';
import { animate, state, style, transition, trigger } from '@angular/animations';
export class ModalOverlayComponent {
    /**
     * @param {?} activeModal
     */
    constructor(activeModal) {
        this.activeModal = activeModal;
        this._ignoreEscapeKey = false;
    }
    /**
     * @return {?}
     */
    _fadeInOut() {
        return state;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _escapeKey(event) {
        if (!this._ignoreEscapeKey) {
            this.activeModal.dismiss();
        }
    }
}
ModalOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-modal-overlay',
                template: '',
                encapsulation: ViewEncapsulation.None,
                // tslint:disable-next-line: no-host-metadata-property
                host: { class: 'hc-modal-overlay' },
                animations: [
                    trigger('fadeInOut', [
                        state('in', style({ opacity: 0.5 })),
                        transition('void <=> *', [
                            style({
                                opacity: 0
                            }),
                            animate('0.2s ease-in-out')
                        ])
                    ])
                ],
                styles: [`
            .hc-modal-overlay {
                background-color: #000;
                position: fixed;
                top: 0;
                left: 0;
                right: 0;
                bottom: 0;
                opacity: 0.5;
                display: block;
            }
        `]
            }] }
];
/** @nocollapse */
ModalOverlayComponent.ctorParameters = () => [
    { type: ActiveModal }
];
ModalOverlayComponent.propDecorators = {
    _ignoreEscapeKey: [{ type: Input }],
    _fadeInOut: [{ type: HostBinding, args: ['@fadeInOut',] }],
    _escapeKey: [{ type: HostListener, args: ['document:keyup.escape', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    ModalOverlayComponent.prototype._ignoreEscapeKey;
    /**
     * @type {?}
     * @private
     */
    ModalOverlayComponent.prototype.activeModal;
}
