// @healthcatalyst/cashmere/esm5/lib/navbar/navbar.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/navbar/navbar.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ContentChildren, ElementRef, HostListener, Input, QueryList, ViewChild, ViewEncapsulation } from '@angular/core';
import { HcPopoverAnchorDirective } from '../pop/directives/popover-anchor.directive';
import { NavbarLinkComponent } from './navbar-link/navbar-link.component';
import { NavbarMobileMenuComponent } from './navbar-mobile-menu/navbar-mobile-menu.component';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
/**
 * The navbar is a wrapper that positions branding, navigation, and other elements in a concise header.
 */
var NavbarComponent = /** @class */ (function () {
    function NavbarComponent(ref) {
        this.ref = ref;
        /**
         * Display name of current user
         */
        this.user = '';
        /**
         * Url to application logo image file
         */
        this.appIcon = '';
        /**
         * Either url to brand icon image file or HcIcon object for a font glyph
         */
        this.brandIcon = '';
        /**
         * Router link triggered when home icon is clicked
         */
        this.homeUri = '';
        /**
         * Fixes the position of navbar to the top of the page. *Default is false.*
         */
        this.fixedTop = false;
        this.unsubscribe$ = new Subject();
        this._menuOpen = false;
        this._linkWidths = [];
        this._linksTotalWidth = 0;
        this._collapse = false;
        this._moreList = [];
    }
    /**
     * @return {?}
     */
    NavbarComponent.prototype._navResize = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._navbarMore) {
            this._navbarMore.closePopover();
        }
        this._moreList = [];
        this._collapse = false;
        // If links is zero the page is smaller than the first responsive breakpoint
        if (this.navbarContent.nativeElement.clientWidth <= 0) {
            return;
        }
        /** @type {?} */
        var linksContainerWidth = this.navContent.nativeElement.offsetWidth;
        /** @type {?} */
        var curLinks = 0;
        // Step through the links until we hit the end of the container, then collapse the
        // remaining into a more menu
        this._navLinks.forEach((/**
         * @param {?} t
         * @param {?} i
         * @return {?}
         */
        function (t, i) {
            curLinks += _this._linkWidths[i];
            /** @type {?} */
            var moreWidth = _this._linksTotalWidth > linksContainerWidth ? 116 : 0;
            if (curLinks + moreWidth < linksContainerWidth) {
                t.show();
            }
            else {
                t.hide();
                _this._collapse = true;
                _this._moreList.push({ name: t.linkText, uri: t.uri });
            }
        }));
        this.ref.detectChanges();
    };
    /** Forces a recalculation of the navbar links to determine how many should be rolling into a More menu.
     * Call this if you've updated the contents of any navbar links. */
    /**
     * Forces a recalculation of the navbar links to determine how many should be rolling into a More menu.
     * Call this if you've updated the contents of any navbar links.
     * @return {?}
     */
    NavbarComponent.prototype.refreshNavLinks = /**
     * Forces a recalculation of the navbar links to determine how many should be rolling into a More menu.
     * Call this if you've updated the contents of any navbar links.
     * @return {?}
     */
    function () {
        this._collectNavLinkWidths();
        this._navResize();
    };
    /**
     * @private
     * @return {?}
     */
    NavbarComponent.prototype._collectNavLinkWidths = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._linkWidths = [];
        this._linksTotalWidth = 0;
        this._navLinks.forEach((/**
         * @param {?} t
         * @return {?}
         */
        function (t) {
            /** @type {?} */
            var isHidden = t._hidden;
            t.show();
            _this._linksTotalWidth += t._getWidth();
            _this._linkWidths.push(t._getWidth());
            if (isHidden) {
                t.hide();
            }
        }));
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.refreshNavLinks();
        }), 100);
        // If links are added dynamically, recheck the navbar link sizing
        this._navLinks.changes.pipe(takeUntil(this.unsubscribe$)).subscribe((/**
         * @return {?}
         */
        function () { return _this.refreshNavLinks(); }));
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype._toggleMobileMenu = /**
     * @return {?}
     */
    function () {
        if (this._menuOpen) {
            this._mobileMenu.first.hide();
            this._menuOpen = false;
        }
        else {
            this._mobileMenu.first.show();
            this._menuOpen = true;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NavbarComponent.prototype._menuClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var clickTarget = event.target.outerHTML;
        // Verify that the click in the mobile menu came from a navigation item
        if (clickTarget.indexOf('hclistline') >= 0 && clickTarget.indexOf('menu-dropdown') === -1) {
            this._toggleMobileMenu();
        }
    };
    Object.defineProperty(NavbarComponent.prototype, "_mobileMenuIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._menuOpen ? 'fa-times' : 'fa-bars';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NavbarComponent.prototype._moreClick = /**
     * @return {?}
     */
    function () {
        if (this._navbarMore) {
            this._navbarMore.closePopover();
        }
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype._brandIconType = /**
     * @return {?}
     */
    function () {
        return typeof this.brandIcon;
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype._brandIconSet = /**
     * @return {?}
     */
    function () {
        if (this.brandIcon && typeof this.brandIcon !== 'string') {
            return this.brandIcon.fontSet;
        }
        else {
            return '';
        }
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype._brandIconGlyph = /**
     * @return {?}
     */
    function () {
        if (this.brandIcon && typeof this.brandIcon !== 'string') {
            return this.brandIcon.fontIcon;
        }
        else {
            return '';
        }
    };
    /**
     * @return {?}
     */
    NavbarComponent.prototype._brandIconSize = /**
     * @return {?}
     */
    function () {
        if (this.brandIcon && typeof this.brandIcon !== 'string' && this.brandIcon.fontSize) {
            return this.brandIcon.fontSize + 'px';
        }
        return '37px';
    };
    NavbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-navbar',
                    template: "<nav #navbar class=\"hc-navbar\" [ngClass]=\"{'fixed-top': fixedTop}\">\n    <div class=\"navbar-brand\">\n        <a [routerLink]=\"homeUri\" class=\"brand\">\n            <img *ngIf=\"_brandIconType() == 'string'\" src=\"{{brandIcon}}\">\n            <hc-icon\n                *ngIf=\"_brandIconType() != 'string'\"\n                [fontSet]=\"_brandIconSet()\"\n                [fontIcon]=\"_brandIconGlyph()\"\n                [ngStyle]=\"{height: _brandIconSize(), 'font-size': _brandIconSize()}\">\n            </hc-icon>\n        </a>\n    </div>\n    <div class=\"navbar-app\">\n        <a [routerLink]=\"homeUri\" class=\"app\" *ngIf=\"appIcon\">\n            <img src=\"{{appIcon}}\">\n        </a>\n    </div>\n    <div #navlinks class=\"hc-navbar-link-container\">\n        <ng-content select=\"hc-navbar-link\"></ng-content>\n        <div class=\"hc-navbar-more-links\" title=\"More\" *ngIf=\"_collapse\" [hcPop]=\"navbarMore\" #moreLink=\"hcPopAnchor\">\n            More<hc-icon fontSet=\"fa\" fontIcon=\"fa-chevron-down\" hcIconSm></hc-icon>\n        </div>\n    </div>\n\n    <div #rightcontainer class=\"hc-navbar-right-container\">\n        <ng-content></ng-content>\n    </div>\n    <div class=\"hc-navbar-mobile-menu\" *ngIf=\"_mobileMenu.first\" (click)=\"_toggleMobileMenu()\">\n        <hc-icon fontSet=\"fa\" [fontIcon]=\"_mobileMenuIcon\" hcIconMd></hc-icon>\n    </div>\n</nav>\n<div (click)=\"_menuClick($event)\">\n    <ng-content select=\"hc-navbar-mobile-menu\"></ng-content>\n</div>\n\n<hc-pop #navbarMore [autoCloseOnContentClick]=\"true\" [showArrow]=\"false\" horizontalAlign=\"end\">\n    <div hcMenu>\n        <ng-container *ngFor=\"let item of _moreList; let i = index\">\n            <button hcMenuItem routerLinkActive=\"active\" [routerLink]=\"item.uri\" (click)=\"_moreClick()\">\n                <span hcMenuText>{{item.name}}</span>\n            </button>\n            <div *ngIf=\"i != _moreList.length - 1\" hcMenuItem hcDivider></div>\n        </ng-container>\n    </div>\n</hc-pop>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".hc-navbar{height:53px;background-color:#384655;display:-webkit-box;display:flex;font-size:15px;line-height:1}.hc-navbar.fixed-top{top:0;position:fixed;right:0;left:0;z-index:970;box-shadow:0 2px 6px rgba(0,0,0,.15)}.hc-navbar .navbar-brand{background-color:#00aeff;min-width:55px;width:55px;flex-basis:auto;height:100%;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center}.hc-navbar .navbar-brand .brand{-webkit-transition:background-color .25s;transition:background-color .25s;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;height:100%;cursor:default}.hc-navbar .navbar-brand .brand>img{height:37px;width:auto}.hc-navbar .navbar-brand .brand>hc-icon{width:100%;color:#fff}.hc-navbar .navbar-app{height:100%;padding:0 50px 0 25px}.hc-navbar .navbar-app a{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center;height:100%}.hc-navbar .navbar-app a img{height:30px;width:auto;padding-top:2px}@media (max-width:576px){.hc-navbar .navbar-brand{display:none}.hc-navbar .navbar-app{padding:0 0 0 5%}}.hc-navbar .hc-navbar-link-container{display:-webkit-box;display:flex;list-style-type:none;margin:0;padding:0;overflow:hidden;width:100%}@media (max-width:768px){.hc-navbar .hc-navbar-link-container{display:none}}.hc-navbar .navbar-link{display:inline-block;height:100%;box-sizing:border-box;padding:22px 30px 0;text-align:center;white-space:nowrap;color:#ccc;border-bottom:5px solid transparent;-webkit-transition:background-color .25s;transition:background-color .25s}.hc-navbar .navbar-link.inactive:hover,.hc-navbar .navbar-link:hover:not(.active){outline:0;color:#fff;background-color:#2e3946}.hc-navbar .navbar-link.active:not(.inactive),.hc-navbar .navbar-link:active{color:#fff;font-weight:600;border-bottom:5px solid #00aeff}.hc-navbar .navbar-link:before{display:block;content:attr(title);font-weight:600;height:1px;color:transparent;overflow:hidden;visibility:hidden;margin-bottom:-1px}.hc-navbar .hc-navbar-right-container{display:-webkit-box;display:flex;-webkit-box-flex:0;flex:0 0 auto;-webkit-box-align:center;align-items:center;position:relative;margin-left:auto}@media (max-width:768px){.hc-navbar .hc-navbar-right-container{display:none}}.hc-navbar .hc-navbar-icon{height:53px;color:#fff;cursor:pointer;-webkit-transition:background-color .25s;transition:background-color .25s;min-width:53px!important;padding:18px 0}.hc-navbar .hc-navbar-icon.inactive:hover,.hc-navbar .hc-navbar-icon:hover:not(.active){outline:0;color:#fff;background-color:#2e3946}.hc-navbar .hc-navbar-username{height:53px;color:#fff;cursor:pointer;-webkit-transition:background-color .25s;transition:background-color .25s;padding:2px 15px 0;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;white-space:nowrap}.hc-navbar .hc-navbar-username.inactive:hover,.hc-navbar .hc-navbar-username:hover:not(.active){outline:0;color:#fff;background-color:#2e3946}.hc-navbar .hc-navbar-username>hc-icon{font-size:12pt;padding-top:2px}.hc-navbar .hc-navbar-username-subtext{color:#c0c5cc}.hc-navbar .hc-navbar-vertical-separator{display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.hc-navbar .hc-navbar-vertical-separator::after{content:'|';font-size:18px;color:#fff}.hc-navbar .hc-navbar-more-links{display:inline-block;white-space:nowrap;cursor:pointer;height:100%;box-sizing:border-box;padding:22px 30px 0;text-align:center;color:#ccc;border-bottom:5px solid transparent;-webkit-transition:background-color .25s;transition:background-color .25s}.hc-navbar .hc-navbar-more-links hc-icon{margin-left:5px;vertical-align:top;margin-top:1px;font-size:13px!important}.hc-navbar .hc-navbar-more-links.inactive:hover,.hc-navbar .hc-navbar-more-links:hover:not(.active){outline:0;color:#fff;background-color:#2e3946}.hc-navbar .hc-navbar-more-links:before{display:block;content:attr(title);font-weight:600;height:1px;color:transparent;overflow:hidden;visibility:hidden;margin-bottom:-1px}.hc-navbar .hc-navbar-mobile-menu{background-color:#262f34;color:#d7dde4;width:55px;flex-basis:auto;height:100%;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;margin-left:auto;cursor:pointer;display:none}.hc-navbar .hc-navbar-mobile-menu hc-icon{font-size:25px!important;height:25px!important;width:25px!important}@media (max-width:768px){.hc-navbar .hc-navbar-mobile-menu{display:-webkit-box;display:flex}}"]
                }] }
    ];
    /** @nocollapse */
    NavbarComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    NavbarComponent.propDecorators = {
        user: [{ type: Input }],
        appIcon: [{ type: Input }],
        brandIcon: [{ type: Input }],
        homeUri: [{ type: Input }],
        fixedTop: [{ type: Input }],
        _mobileMenu: [{ type: ContentChildren, args: [NavbarMobileMenuComponent,] }],
        _navLinks: [{ type: ContentChildren, args: [NavbarLinkComponent,] }],
        navbarContent: [{ type: ViewChild, args: ['navbar', { static: false },] }],
        navContent: [{ type: ViewChild, args: ['navlinks', { static: false },] }],
        _navbarMore: [{ type: ViewChild, args: ['moreLink', { static: false },] }],
        _navResize: [{ type: HostListener, args: ['window:resize',] }]
    };
    return NavbarComponent;
}());
export { NavbarComponent };
if (false) {
    /**
     * Display name of current user
     * @type {?}
     */
    NavbarComponent.prototype.user;
    /**
     * Url to application logo image file
     * @type {?}
     */
    NavbarComponent.prototype.appIcon;
    /**
     * Either url to brand icon image file or HcIcon object for a font glyph
     * @type {?}
     */
    NavbarComponent.prototype.brandIcon;
    /**
     * Router link triggered when home icon is clicked
     * @type {?}
     */
    NavbarComponent.prototype.homeUri;
    /**
     * Fixes the position of navbar to the top of the page. *Default is false.*
     * @type {?}
     */
    NavbarComponent.prototype.fixedTop;
    /** @type {?} */
    NavbarComponent.prototype._mobileMenu;
    /** @type {?} */
    NavbarComponent.prototype._navLinks;
    /** @type {?} */
    NavbarComponent.prototype.navbarContent;
    /** @type {?} */
    NavbarComponent.prototype.navContent;
    /** @type {?} */
    NavbarComponent.prototype._navbarMore;
    /**
     * @type {?}
     * @private
     */
    NavbarComponent.prototype.unsubscribe$;
    /**
     * @type {?}
     * @private
     */
    NavbarComponent.prototype._menuOpen;
    /**
     * @type {?}
     * @private
     */
    NavbarComponent.prototype._linkWidths;
    /**
     * @type {?}
     * @private
     */
    NavbarComponent.prototype._linksTotalWidth;
    /** @type {?} */
    NavbarComponent.prototype._collapse;
    /** @type {?} */
    NavbarComponent.prototype._moreList;
    /**
     * @type {?}
     * @private
     */
    NavbarComponent.prototype.ref;
}
