// @healthcatalyst/cashmere/esm5/lib/navbar/navbar-mobile-menu/navbar-mobile-menu.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/navbar/navbar-mobile-menu/navbar-mobile-menu.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation } from '@angular/core';
/**
 * Navigation dropdown for small screen sizes
 */
var NavbarMobileMenuComponent = /** @class */ (function () {
    function NavbarMobileMenuComponent() {
        this._yPos = '-100';
    }
    /** Show the component from view */
    /**
     * Show the component from view
     * @return {?}
     */
    NavbarMobileMenuComponent.prototype.show = /**
     * Show the component from view
     * @return {?}
     */
    function () {
        this._yPos = '0';
    };
    /** Hide the component from view */
    /**
     * Hide the component from view
     * @return {?}
     */
    NavbarMobileMenuComponent.prototype.hide = /**
     * Hide the component from view
     * @return {?}
     */
    function () {
        this._yPos = '-100';
    };
    NavbarMobileMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-navbar-mobile-menu',
                    template: "<div class=\"hc-navbar-mobile-menu-dropdown\" [style.top.%]=\"_yPos\">\n    <ng-content></ng-content>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".hc-navbar-mobile-menu-dropdown{background-color:#262f34;display:none;margin-top:53px;padding-top:15px;position:fixed;-webkit-transition:top .7s;transition:top .7s;width:100%;z-index:960}@media (max-width:768px){.hc-navbar-mobile-menu-dropdown{display:inline}}.hc-navbar-mobile-menu-dropdown hr{background-color:#51646f;border:0;height:1px;margin:0 0 0 5%!important}.hc-navbar-mobile-menu-dropdown hc-list a{text-decoration:none}.hc-navbar-mobile-menu-dropdown hc-list .hc-list-item{color:#d7dde4;cursor:pointer;display:block;padding-left:2%;-webkit-transition:background-color .25s;transition:background-color .25s}.hc-navbar-mobile-menu-dropdown hc-list .hc-list-item:hover{background-color:#1b2225}.hc-navbar-mobile-menu-dropdown hc-list .hc-list-item:focus{outline:0}.hc-navbar-mobile-menu-dropdown hc-list .active-link{background-color:#1b2225;border-left:8px solid #00aeff;color:#fff;font-weight:700}.hc-navbar-mobile-menu-dropdown hc-list .hc-list-item.active-link .hc-list-content{padding-left:4px}"]
                }] }
    ];
    /** @nocollapse */
    NavbarMobileMenuComponent.ctorParameters = function () { return []; };
    return NavbarMobileMenuComponent;
}());
export { NavbarMobileMenuComponent };
if (false) {
    /** @type {?} */
    NavbarMobileMenuComponent.prototype._yPos;
}
