// @healthcatalyst/cashmere/esm5/lib/datepicker/datepicker-content/datepicker-content.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/datepicker/datepicker-content/datepicker-content.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, ChangeDetectionStrategy, ViewChild } from '@angular/core';
import { HcDatepickerAnimations } from '../datepicker-animations';
import { CalendarComponent } from '../calendar/calendar.component';
/**
 * Component used as the content for the datepicker popup. We use this instead of using
 * hcCalendar directly as the content so we can control the initial focus. This also gives us a
 * place to put additional features of the popup that are not part of the calendar itself in the
 * future. (e.g. confirmation buttons).
 * \@docs-private
 */
var DatepickerContentComponent = /** @class */ (function () {
    function DatepickerContentComponent() {
    }
    /**
     * @return {?}
     */
    DatepickerContentComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this._calendar.focusActiveCell();
    };
    /** Close the datepicker automatically on selection only if in date mode */
    /**
     * Close the datepicker automatically on selection only if in date mode
     * @return {?}
     */
    DatepickerContentComponent.prototype.autoClose = /**
     * Close the datepicker automatically on selection only if in date mode
     * @return {?}
     */
    function () {
        if (this.datepicker.mode === 'date') {
            this.datepicker.close();
        }
    };
    DatepickerContentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-datepicker-content',
                    template: "<hc-calendar cdkTrapFocus\n    [id]=\"datepicker.id\"\n    [ngClass]=\"datepicker.panelClass\"\n    [mode]=\"datepicker.mode\"\n    [hourCycle]=\"datepicker.hourCycle\"\n    [startAt]=\"datepicker.startAt\"\n    [startView]=\"datepicker.startView\"\n    [minDate]=\"datepicker._minDate\"\n    [maxDate]=\"datepicker._maxDate\"\n    [dateFilter]=\"datepicker._dateFilter\"\n    [headerComponent]=\"datepicker.calendarHeaderComponent\"\n    [selected]=\"datepicker._selected\"\n    [dateClass]=\"datepicker.dateClass\"\n    [@fadeInCalendar]=\"'enter'\"\n    (selectedChange)=\"datepicker.select($event)\"\n    (yearSelected)=\"datepicker._selectYear($event)\"\n    (monthSelected)=\"datepicker._selectMonth($event)\"\n    (_userSelection)=\"autoClose()\">\n</hc-calendar>\n",
                    // tslint:disable-next-line:no-host-metadata-property
                    host: {
                        class: 'hc-datepicker-content',
                        '[@transformPanel]': '"enter"'
                    },
                    animations: [HcDatepickerAnimations.transformPanel, HcDatepickerAnimations.fadeInCalendar],
                    exportAs: 'hcDatepickerContent',
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    DatepickerContentComponent.propDecorators = {
        _calendar: [{ type: ViewChild, args: [CalendarComponent, { static: false },] }]
    };
    return DatepickerContentComponent;
}());
export { DatepickerContentComponent };
if (false) {
    /**
     * Reference to the internal calendar component.
     * @type {?}
     */
    DatepickerContentComponent.prototype._calendar;
    /**
     * Reference to the datepicker that created the overlay.
     * @type {?}
     */
    DatepickerContentComponent.prototype.datepicker;
    /**
     * Whether the datepicker is above or below the input.
     * @type {?}
     */
    DatepickerContentComponent.prototype._isAbove;
}
