// @healthcatalyst/cashmere/esm2015/lib/date-range/services/config-store.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-range/services/config-store.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';
export class ConfigStoreService {
    constructor() {
        this.defaultOptions = {
            presets: [],
            format: 'medium',
            mode: 'date',
            hourCycle: 12,
            excludeWeekends: false,
            locale: 'en-us',
            applyLabel: 'Apply',
            cancelLabel: 'Cancel',
            startDatePrefix: 'Start date:',
            endDatePrefix: 'End date:',
            invalidDateLabel: 'Please enter valid date'
        };
        this.dateRangeOptionsSubject = new BehaviorSubject(this.defaultOptions);
        this.rangeUpdateSubject = new BehaviorSubject({ fromDate: undefined, toDate: undefined });
        this.presetUpdateSubject = new BehaviorSubject({
            fromDate: undefined,
            toDate: undefined
        });
        this.weekendFilter = (/**
         * @return {?}
         */
        () => true);
        this.emptyWeekendFilter = (/**
         * @return {?}
         */
        () => true);
        this.excludeWeekendFilter = (/**
         * @param {?} d
         * @return {?}
         */
        (d) => {
            /** @type {?} */
            const day = d.getDay();
            return day !== 0 && day !== 6;
        });
        this.dateRangeOptions$ = this.dateRangeOptionsSubject.pipe(map((/**
         * @param {?} options
         * @return {?}
         */
        (options) => {
            return Object.assign({}, this.defaultOptions, options);
        })), tap((/**
         * @param {?} options
         * @return {?}
         */
        (options) => {
            if (!!options.excludeWeekends) {
                this.weekendFilter = this.excludeWeekendFilter;
            }
            else {
                this.weekendFilter = this.emptyWeekendFilter;
            }
        })));
        this.rangeUpdate$ = this.rangeUpdateSubject.pipe();
        this.presetUpdate$ = this.presetUpdateSubject.pipe();
    }
    /**
     * @param {?} options
     * @return {?}
     */
    updateDateRangeOptions(options) {
        this.dateRangeOptionsSubject.next(options);
    }
    /**
     * @param {?} dateRange
     * @return {?}
     */
    updateRange(dateRange) {
        this.rangeUpdateSubject.next(dateRange);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    updatePreset(value) {
        this.presetUpdateSubject.next(value);
    }
}
ConfigStoreService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ConfigStoreService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.defaultOptions;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.dateRangeOptionsSubject;
    /** @type {?} */
    ConfigStoreService.prototype.dateRangeOptions$;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.rangeUpdateSubject;
    /** @type {?} */
    ConfigStoreService.prototype.rangeUpdate$;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.presetUpdateSubject;
    /** @type {?} */
    ConfigStoreService.prototype.presetUpdate$;
    /** @type {?} */
    ConfigStoreService.prototype.weekendFilter;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.emptyWeekendFilter;
    /**
     * @type {?}
     * @private
     */
    ConfigStoreService.prototype.excludeWeekendFilter;
}
