// @healthcatalyst/cashmere/esm5/lib/scroll-nav/nav/scroll-nav.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/scroll-nav/nav/scroll-nav.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewEncapsulation, QueryList, ContentChildren } from '@angular/core';
import { ScrollNavLinkDirective } from './scroll-nav-link.directive';
/**
 * Container for scroll navigation links.
 */
var HcScrollNavComponent = /** @class */ (function () {
    function HcScrollNavComponent(_elementRef) {
        this._elementRef = _elementRef;
        this.ACTIVE_CLASS = 'hc-scroll-nav-link-active';
    }
    Object.defineProperty(HcScrollNavComponent.prototype, "_links", {
        get: /**
         * @return {?}
         */
        function () {
            return this.linkList.toArray().map((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e._el.nativeElement; }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HcScrollNavComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this._links && this._links[0]) {
            this._links[0].classList.add(this.ACTIVE_CLASS);
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    HcScrollNavComponent.prototype._setActiveClassById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var link = this.linkList.find((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return e.hcScrollLink === id; }));
        if (!link) {
            throw new Error("Failed to mark active class. Could not find the element with the data target for id: " + id + ".");
        }
        this.setActiveClass(link._el.nativeElement);
    };
    /**
     * @private
     * @param {?} element
     * @return {?}
     */
    HcScrollNavComponent.prototype.setActiveClass = /**
     * @private
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _this = this;
        this._links.forEach((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.classList.remove(_this.ACTIVE_CLASS);
        }));
        element.classList.add(this.ACTIVE_CLASS);
    };
    HcScrollNavComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-scroll-nav',
                    encapsulation: ViewEncapsulation.None,
                    template: "<div class=\"hc-scroll-nav-links-container\">\n  <ng-content></ng-content>\n</div>",
                    styles: [".hc-scroll-nav-links-container{margin-right:30px}.hc-scroll-nav-links-container ul{list-style:none}.hc-scroll-nav-link{font-size:1rem;background-color:transparent;color:#333;cursor:pointer;font-weight:500;padding:10px 15px}.hc-scroll-nav-link:hover{background-color:rgba(215,221,228,.25)}.hc-scroll-nav-link:active{background-color:rgba(215,221,228,.5)}.hc-scroll-nav-link-active,.hc-scroll-nav-link-active:hover{background-color:#00aeff;color:#fff;font-weight:500}"]
                }] }
    ];
    /** @nocollapse */
    HcScrollNavComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    HcScrollNavComponent.propDecorators = {
        linkList: [{ type: ContentChildren, args: [ScrollNavLinkDirective,] }]
    };
    return HcScrollNavComponent;
}());
export { HcScrollNavComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcScrollNavComponent.prototype.linkList;
    /**
     * @type {?}
     * @private
     */
    HcScrollNavComponent.prototype.ACTIVE_CLASS;
    /** @type {?} */
    HcScrollNavComponent.prototype._elementRef;
}
