// @healthcatalyst/cashmere/esm5/lib/progress-indicators/progress-dots.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/progress-indicators/progress-dots.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
/** @type {?} */
var supportedColors = ['light', 'dark'];
/**
 * @param {?} inputStr
 * @return {?}
 */
export function validateColorInput(inputStr) {
    if (supportedColors.indexOf(inputStr) < 0) {
        throw Error('Unsupported progress dots color value: ' + inputStr);
    }
}
var ProgressDotsComponent = /** @class */ (function () {
    function ProgressDotsComponent() {
        this._color = 'dark';
        /**
         * If true, the loader will center itself within its container.
         */
        this.isCentered = true;
        /**
         * If true, you'll get a teeny tiny little loader.
         */
        this.isMini = false;
    }
    Object.defineProperty(ProgressDotsComponent.prototype, "color", {
        /** Use `light` on darker backgrounds and `dark` for lighter backgrounds. */
        get: /**
         * Use `light` on darker backgrounds and `dark` for lighter backgrounds.
         * @return {?}
         */
        function () {
            return this._color;
        },
        set: /**
         * @param {?} colorVal
         * @return {?}
         */
        function (colorVal) {
            validateColorInput(colorVal);
            this._color = colorVal;
        },
        enumerable: true,
        configurable: true
    });
    ProgressDotsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-progress-dots',
                    template: "<div class=\"hc-dots-container\" [ngClass]=\"{'center-dots': isCentered }\">\n    <div class=\"hc-dots-loader\" [ngClass]=\"{'loader-light': color === 'light', 'loader-mini': isMini}\">\n        <div class=\"loader-animate\"></div>\n    </div>\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    ProgressDotsComponent.propDecorators = {
        color: [{ type: Input }],
        isCentered: [{ type: Input }],
        isMini: [{ type: Input }]
    };
    return ProgressDotsComponent;
}());
export { ProgressDotsComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProgressDotsComponent.prototype._color;
    /**
     * If true, the loader will center itself within its container.
     * @type {?}
     */
    ProgressDotsComponent.prototype.isCentered;
    /**
     * If true, you'll get a teeny tiny little loader.
     * @type {?}
     */
    ProgressDotsComponent.prototype.isMini;
}
