// @healthcatalyst/cashmere/esm5/lib/toaster/hc-toaster.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/toaster/hc-toaster.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HcToastComponent } from './hc-toast.component';
import { HcToasterService } from './hc-toaster.service';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PortalModule } from '@angular/cdk/portal';
import { OverlayModule } from '@angular/cdk/overlay';
var ToasterModule = /** @class */ (function () {
    function ToasterModule() {
    }
    ToasterModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, PortalModule, OverlayModule],
                    exports: [HcToastComponent],
                    declarations: [HcToastComponent],
                    providers: [HcToasterService],
                    entryComponents: [HcToastComponent]
                },] }
    ];
    return ToasterModule;
}());
export { ToasterModule };
