// @healthcatalyst/cashmere/esm2015/lib/picklist/picklist.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/picklist/picklist.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { PicklistPaneComponent } from './pane/picklist-pane.component';
import { PicklistComponent } from './picklist.component';
import { ButtonModule } from '../button/button.module';
import { IconModule } from '../icon/icon.module';
import { InputModule } from '../input/input.module';
import { FormFieldModule } from '../form-field/hc-form-field.module';
import { ProgressIndicatorsModule } from '../progress-indicators/progress-indicators.module';
export { PicklistPaneComponent } from './pane/picklist-pane.component';
export { PicklistComponent } from './picklist.component';
export { PicklistRemoteQueryOptions } from './picklist.model';
export class PicklistModule {
}
PicklistModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, FormsModule, ButtonModule, IconModule, InputModule, FormFieldModule, ProgressIndicatorsModule],
                declarations: [PicklistPaneComponent, PicklistComponent],
                exports: [PicklistPaneComponent, PicklistComponent]
            },] }
];
