// @healthcatalyst/cashmere/esm2015/lib/table/row.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/table/row.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// tslint:disable:directive-class-suffix use-input-property-decorator component-class-suffix
// tslint:disable:no-host-metadata-property no-inputs-metadata-property
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { ChangeDetectionStrategy, Component, Directive, ViewEncapsulation } from '@angular/core';
import { CDK_ROW_TEMPLATE, CdkFooterRow, CdkFooterRowDef, CdkHeaderRow, CdkHeaderRowDef, CdkRow, CdkRowDef } from '@angular/cdk/table';
/**
 * Header row definition for the hc-table.
 * Captures the header row's template and other header properties such as the columns to display.
 */
export class HcHeaderRowDef extends CdkHeaderRowDef {
}
HcHeaderRowDef.decorators = [
    { type: Directive, args: [{
                selector: '[hcHeaderRowDef]',
                providers: [{ provide: CdkHeaderRowDef, useExisting: HcHeaderRowDef }],
                inputs: ['columns: hcHeaderRowDef', 'sticky: hcHeaderRowDefSticky']
            },] }
];
/**
 * Footer row definition for the mat-table.
 * Captures the footer row's template and other footer properties such as the columns to display.
 */
export class HcFooterRowDef extends CdkFooterRowDef {
}
HcFooterRowDef.decorators = [
    { type: Directive, args: [{
                selector: '[hcFooterRowDef]',
                providers: [{ provide: CdkFooterRowDef, useExisting: HcFooterRowDef }],
                inputs: ['columns: hcFooterRowDef', 'sticky: hcFooterRowDefSticky']
            },] }
];
/**
 * Data row definition for the mat-table.
 * Captures the footer row's template and other footer properties such as the columns to display and
 * a when predicate that describes when this row should be used.
 * @template T
 */
export class HcRowDef extends CdkRowDef {
}
HcRowDef.decorators = [
    { type: Directive, args: [{
                selector: '[hcRowDef]',
                providers: [{ provide: CdkRowDef, useExisting: HcRowDef }],
                inputs: ['columns: hcRowDefColumns', 'when: hcRowDefWhen']
            },] }
];
/**
 * Footer template container that contains the cell outlet. Adds the right class and role.
 */
export class HcHeaderRow extends CdkHeaderRow {
}
HcHeaderRow.decorators = [
    { type: Component, args: [{
                selector: 'hc-header-row, tr[hc-header-row]',
                template: CDK_ROW_TEMPLATE,
                host: {
                    class: 'hc-header-row',
                    role: 'row'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                exportAs: 'hcHeaderRow',
                providers: [{ provide: CdkHeaderRow, useExisting: HcHeaderRow }]
            }] }
];
/**
 * Footer template container that contains the cell outlet. Adds the right class and role.
 */
export class HcFooterRow extends CdkFooterRow {
}
HcFooterRow.decorators = [
    { type: Component, args: [{
                selector: 'hc-footer-row, tr[hc-footer-row]',
                template: CDK_ROW_TEMPLATE,
                host: {
                    class: 'hc-footer-row',
                    role: 'row'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                exportAs: 'hcFooterRow',
                providers: [{ provide: CdkFooterRow, useExisting: HcFooterRow }]
            }] }
];
/**
 * Data row template container that contains the cell outlet. Adds the right class and role.
 */
export class HcRow extends CdkRow {
}
HcRow.decorators = [
    { type: Component, args: [{
                selector: 'hc-row, tr[hc-row]',
                template: CDK_ROW_TEMPLATE,
                host: {
                    class: 'hc-row',
                    role: 'row'
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                exportAs: 'hcRow',
                providers: [{ provide: CdkRow, useExisting: HcRow }]
            }] }
];
