// @healthcatalyst/cashmere/esm2015/lib/table/bind-observable/bind-observable.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/table/bind-observable/bind-observable.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* Copyright 2017 PSanetra <code@psanetra.de>

Open Source typescript decorator which binds class properties to observable companion properties.
https://github.com/PSanetra/bind-observable

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */
import { ReplaySubject } from 'rxjs';
import { isBindObservableOpts } from './bind-observable-options';
/** @type {?} */
const subjects = new WeakMap();
/** @type {?} */
const values = new WeakMap();
/**
 * @param {?} instance
 * @param {?} key
 * @return {?}
 */
function subject(instance, key) {
    /** @type {?} */
    let subjectByProp = subjects.get(instance);
    if (!subjectByProp) {
        subjectByProp = new Map();
        subjects.set(instance, subjectByProp);
    }
    /** @type {?} */
    let _subject = subjectByProp.get(key);
    if (!_subject) {
        _subject = new ReplaySubject(1);
        subjectByProp.set(key, _subject);
    }
    return _subject;
}
/**
 * @param {?} instance
 * @return {?}
 */
function valueMap(instance) {
    /** @type {?} */
    let _valueMap = values.get(instance);
    if (!_valueMap) {
        _valueMap = new Map();
        values.set(instance, _valueMap);
    }
    return _valueMap;
}
/**
 * @param {?} target
 * @param {?} observableKey
 * @return {?}
 */
function defineObservableProperty(target, observableKey) {
    Object.defineProperty(target, observableKey, {
        configurable: true,
        enumerable: false,
        /**
         * @return {?}
         */
        get() {
            return subject(this, observableKey);
        }
    });
}
/**
 * @param {?} target
 * @param {?} propertyKey
 * @param {?} observableKey
 * @return {?}
 */
function redefineSimpleProperty(target, propertyKey, observableKey) {
    Object.defineProperty(target, propertyKey, {
        configurable: true,
        enumerable: true,
        /**
         * @param {?} value
         * @return {?}
         */
        set(value) {
            valueMap(this).set(propertyKey, value);
            subject(this, observableKey).next(value);
        },
        /**
         * @return {?}
         */
        get() {
            return valueMap(this).get(propertyKey);
        }
    });
}
/**
 * @param {?} target
 * @param {?} propertyKey
 * @param {?} observableKey
 * @param {?} emitRawSetterValue
 * @param {?} descriptor
 * @return {?}
 */
function redefineAccessorProperty(target, propertyKey, observableKey, emitRawSetterValue, descriptor) {
    Object.defineProperty(target, propertyKey, {
        configurable: descriptor.configurable,
        enumerable: descriptor.enumerable,
        /**
         * @param {?} value
         * @return {?}
         */
        set(value) {
            if (!descriptor.set) {
                throw new Error(`Property ${propertyKey} doesn't have a setter and cannot be written`);
            }
            descriptor.set.call(this, value);
            /** @type {?} */
            const companionProp = subject(this, observableKey);
            if (emitRawSetterValue || !descriptor.get) {
                companionProp.next(value);
            }
            else {
                companionProp.next(descriptor.get());
            }
        },
        /**
         * @return {?}
         */
        get() {
            if (!descriptor.get) {
                throw new Error(`Property ${propertyKey} doesn't have a getter and cannot be read`);
            }
            return descriptor.get.call(this);
        }
    });
}
/**
 * Binds a property to an observable companion property.
 * The observable companion property will emit on all assignments (including initialization),
 * but will not emit undefined if undefined is not explicitly assigned on initialization.
 * @param {?=} observableKeyOrOpts
 * @return {?}
 */
export function BindObservable(observableKeyOrOpts) {
    return (/**
     * @param {?} target
     * @param {?} propertyKey
     * @return {?}
     */
    (target, propertyKey) => {
        /** @type {?} */
        const opts = isBindObservableOpts(observableKeyOrOpts) ? observableKeyOrOpts : {};
        if (typeof observableKeyOrOpts === 'string') {
            opts.key = observableKeyOrOpts;
        }
        /** @type {?} */
        const observableKey = opts.key || propertyKey + '$';
        // The third parameter of this function (descriptor) is passed only if the decorated property
        // is an accessor, but it won't be passed if another decorator has replaced the descriptor.
        // See Property Decorators at https://www.typescriptlang.org/docs/handbook/decorators.html
        // Because of this, we are forced to retrieve the current descriptor with Reflection API
        /** @type {?} */
        const descriptor = Reflect.getOwnPropertyDescriptor(target, propertyKey);
        delete target[propertyKey];
        delete target[observableKey];
        defineObservableProperty(target, observableKey);
        if (descriptor !== undefined) {
            redefineAccessorProperty(target, propertyKey, observableKey, !!opts.emitRawSetterValue, descriptor);
        }
        else {
            redefineSimpleProperty(target, propertyKey, observableKey);
        }
    });
}
