// @healthcatalyst/cashmere/esm2015/lib/pipes/number-abbreviator/number-abbreviator.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/number-abbreviator/number-abbreviator.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from "@angular/core";
/**
 * Creates an abbreviated number for display. 1,234,234 => 1.23M
 */
export class NumberAbbreviatorPipe {
    /**
     * @param {?} value the value to be abbreviated
     * @param {?=} decimalPoints how many decimal points to round to
     * @param {?=} threshold represents the number at which we should start abbreviating;
     * if the threshold is 1 million, we'll leave number less than 1 million as is (aside from adding commas or decimal points)
     * @return {?}
     */
    transform(value, decimalPoints = 2, threshold = 1000000) {
        if (value === null) {
            return value;
        }
        // use defaults if given funky parameters
        if (decimalPoints < 0 || !Number.isInteger(decimalPoints)) {
            decimalPoints = 2;
        }
        if (threshold < 1000 || !Number.isInteger(threshold)) {
            threshold = 1000;
        }
        /** @type {?} */
        const abs = Math.abs(value);
        /** @type {?} */
        let key = "";
        /** @type {?} */
        let roundedStringVal = "";
        /** @type {?} */
        const isNegative = value < 0;
        if (abs >= threshold) {
            // Abbreviate, adding unit if greater than specified limit
            for (const power of NumberAbbreviatorPipe.powers) {
                /** @type {?} */
                let reduced = abs / power.value;
                reduced = this.roundValue(reduced, decimalPoints);
                if (reduced >= 1) {
                    roundedStringVal = reduced.toString();
                    key = power.key;
                    break;
                }
            }
        }
        else {
            // Just round decimals and add commas if less than specified limit
            roundedStringVal = this.roundValue(abs, decimalPoints).toLocaleString();
        }
        return (isNegative ? "-" : "") + roundedStringVal + key;
    }
    /**
     * @private
     * @param {?} value
     * @param {?=} decimalPoints
     * @return {?}
     */
    roundValue(value, decimalPoints = 2) {
        /** @type {?} */
        const rounder = Math.pow(10, decimalPoints);
        return Math.round(value * rounder) / rounder;
    }
}
NumberAbbreviatorPipe.powers = [
    { key: "Q", value: Math.pow(10, 15) },
    { key: "T", value: Math.pow(10, 12) },
    { key: "B", value: Math.pow(10, 9) },
    { key: "M", value: Math.pow(10, 6) },
    { key: "K", value: 1000 }
];
NumberAbbreviatorPipe.decorators = [
    { type: Pipe, args: [{ name: "abbreviateNumber" },] }
];
if (false) {
    /** @type {?} */
    NumberAbbreviatorPipe.powers;
}
