// @healthcatalyst/cashmere/esm2015/lib/pipes/ellipsis/ellipsis.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/ellipsis/ellipsis.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
export class EllipsisPipe {
    /**
     * @param {?} value
     * @param {?} length
     * @param {?=} mode
     * @return {?}
     */
    transform(value, length, mode = 'characters') {
        if (typeof value !== 'string' || typeof length !== 'number' || length < 0 || isNaN(length) || Math.floor(length) !== length) {
            return value;
        }
        /** @type {?} */
        const truncated = mode === 'words' ? this.firstNWords(value, length) : this.firstNCharacters(value, length);
        if (truncated === value) {
            return value;
        }
        return truncated + EllipsisPipe.ELLIPSIS;
    }
    /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    firstNWords(value, n) {
        /** @type {?} */
        const words = value.split(/\s+/g);
        if (words.length <= n) {
            return value;
        }
        /** @type {?} */
        const pattern = new RegExp('^' + words.slice(0, n).join('\\s+'));
        return (/** @type {?} */ (pattern.exec(value)))[0];
    }
    /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    firstNCharacters(value, n) {
        return value.length > n ? value.substring(0, n) : value;
    }
}
EllipsisPipe.ELLIPSIS = '…';
EllipsisPipe.decorators = [
    { type: Pipe, args: [{
                name: 'ellipsis',
                pure: true
            },] }
];
if (false) {
    /** @type {?} */
    EllipsisPipe.ELLIPSIS;
}
