// @healthcatalyst/cashmere/esm5/lib/pipes/ellipsis/ellipsis.pipe.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/pipes/ellipsis/ellipsis.pipe.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
var EllipsisPipe = /** @class */ (function () {
    function EllipsisPipe() {
    }
    /**
     * @param {?} value
     * @param {?} length
     * @param {?=} mode
     * @return {?}
     */
    EllipsisPipe.prototype.transform = /**
     * @param {?} value
     * @param {?} length
     * @param {?=} mode
     * @return {?}
     */
    function (value, length, mode) {
        if (mode === void 0) { mode = 'characters'; }
        if (typeof value !== 'string' || typeof length !== 'number' || length < 0 || isNaN(length) || Math.floor(length) !== length) {
            return value;
        }
        /** @type {?} */
        var truncated = mode === 'words' ? this.firstNWords(value, length) : this.firstNCharacters(value, length);
        if (truncated === value) {
            return value;
        }
        return truncated + EllipsisPipe.ELLIPSIS;
    };
    /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    EllipsisPipe.prototype.firstNWords = /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    function (value, n) {
        /** @type {?} */
        var words = value.split(/\s+/g);
        if (words.length <= n) {
            return value;
        }
        /** @type {?} */
        var pattern = new RegExp('^' + words.slice(0, n).join('\\s+'));
        return (/** @type {?} */ (pattern.exec(value)))[0];
    };
    /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    EllipsisPipe.prototype.firstNCharacters = /**
     * @private
     * @param {?} value
     * @param {?} n
     * @return {?}
     */
    function (value, n) {
        return value.length > n ? value.substring(0, n) : value;
    };
    EllipsisPipe.ELLIPSIS = '…';
    EllipsisPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'ellipsis',
                    pure: true
                },] }
    ];
    return EllipsisPipe;
}());
export { EllipsisPipe };
if (false) {
    /** @type {?} */
    EllipsisPipe.ELLIPSIS;
}
