// @healthcatalyst/cashmere/esm2015/lib/checkbox/checkbox-required.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/checkbox/checkbox-required.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* tslint:disable:directive-selector */
import { Directive, forwardRef, HostBinding } from '@angular/core';
import { CheckboxRequiredValidator, NG_VALIDATORS } from '@angular/forms';
/** @type {?} */
export const HC_CHECKBOX_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => HcCheckboxRequiredValidatorDirective)),
    multi: true
};
/**
 * \@docs-private
 */
export class HcCheckboxRequiredValidatorDirective extends CheckboxRequiredValidator {
    /**
     * @return {?}
     */
    get isRequired() {
        return this.required ? '' : null;
    }
}
HcCheckboxRequiredValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `hc-checkbox[required][formControlName],
             hc-checkbox[required][formControl],hc-checkbox[required][ngModel]`,
                providers: [HC_CHECKBOX_REQUIRED_VALIDATOR]
            },] }
];
HcCheckboxRequiredValidatorDirective.propDecorators = {
    isRequired: [{ type: HostBinding, args: ['attr.required',] }]
};
