// @healthcatalyst/cashmere/esm5/lib/form-field/hc-form-field.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-field/hc-form-field.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ContentChildren, HostBinding, QueryList, ViewEncapsulation, Input, ElementRef } from '@angular/core';
import { HcFormControlComponent } from './hc-form-control.component';
import { HcErrorComponent } from './hc-error.component';
import { HcPrefixDirective } from './hc-prefix.directive';
import { HcSuffixDirective } from './hc-suffix.directive';
import { HcLabelComponent } from './hc-label.component';
import { parseBooleanAttribute } from '../util';
import { InputDirective } from '../input/input.directive';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
/**
 * @return {?}
 */
export function getControlMissing() {
    return new Error("HcFormField must contain a component that extends HcFormControl");
}
/**
 * Container for form fields that applies Cashmere styling and behavior
 */
var HcFormFieldComponent = /** @class */ (function () {
    function HcFormFieldComponent(_elementRef) {
        this._elementRef = _elementRef;
        this._inline = false;
        this._tight = false;
        this.unsubscribe$ = new Subject();
        this._classHcFormFieldClass = true;
    }
    Object.defineProperty(HcFormFieldComponent.prototype, "_disabledClass", {
        get: /**
         * @return {?}
         */
        function () {
            if (this._inputChildren.length > 0) {
                return this._inputChildren.first.disabled;
            }
            else {
                return this._control._isDisabled;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HcFormFieldComponent.prototype, "hasLabel", {
        /** Read-only boolean value of whether the form field has an associated label element */
        get: /**
         * Read-only boolean value of whether the form field has an associated label element
         * @return {?}
         */
        function () {
            return !!this._labelChildren.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HcFormFieldComponent.prototype, "hasInput", {
        /** Read-only boolean value of whether the form field has an input element */
        get: /**
         * Read-only boolean value of whether the form field has an input element
         * @return {?}
         */
        function () {
            return !!this._inputChildren.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HcFormFieldComponent.prototype, "inline", {
        /** Whether the form elements should be stacked (default), or inline */
        get: /**
         * Whether the form elements should be stacked (default), or inline
         * @return {?}
         */
        function () {
            return this._inline;
        },
        set: /**
         * @param {?} isInline
         * @return {?}
         */
        function (isInline) {
            this._inline = parseBooleanAttribute(isInline);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HcFormFieldComponent.prototype, "tight", {
        /** If true, condense the default padding on all included elements and reduce the font size. *Defaults to `false`.*  */
        get: /**
         * If true, condense the default padding on all included elements and reduce the font size. *Defaults to `false`.*
         * @return {?}
         */
        function () {
            return this._tight;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._tight = parseBooleanAttribute(value);
            this._updateTightControls();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HcFormFieldComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this._control) {
            throw getControlMissing();
        }
        else {
            this._updateTightControls();
            // Pass the current tight setting to controls that are added dynamically to the FormField
            this._controls.changes.pipe(takeUntil(this.unsubscribe$)).subscribe((/**
             * @return {?}
             */
            function () { return _this._updateTightControls(); }));
        }
    };
    /**
     * @return {?}
     */
    HcFormFieldComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    };
    /**
     * @return {?}
     */
    HcFormFieldComponent.prototype._updateTightControls = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._controls) {
            this._controls.forEach((/**
             * @param {?} control
             * @return {?}
             */
            function (control) {
                control.tight = _this._tight;
            }));
        }
    };
    /**
     * @return {?}
     */
    HcFormFieldComponent.prototype.getConnectedOverlayOrigin = /**
     * @return {?}
     */
    function () {
        return this._elementRef;
    };
    /**
     * @return {?}
     */
    HcFormFieldComponent.prototype._shouldShowErrorMessages = /**
     * @return {?}
     */
    function () {
        return (this._control._errorState &&
            ((this._errorChildren && this._errorChildren.length > 0) ||
                (!!this._control._errorMessage && this._control._errorMessage.length > 0)));
    };
    HcFormFieldComponent.decorators = [
        { type: Component, args: [{
                    selector: 'hc-form-field',
                    template: "<div\n    [class.hc-form-field-wrapper-inline]=\"inline\"\n    [class.hc-form-field-wrapper]=\"!inline\"\n    [class.hc-form-field-no-label]=\"!hasLabel\"\n    [class.hc-form-field-wrapper-tight]=\"tight\" >\n    <span *ngIf=\"hasLabel\" [ngClass]=\"inline ? 'hc-form-field-label-wrapper-inline' : 'hc-form-field-label-wrapper'\" [class.hc-form-field-label-wrapper-tight]=\"tight\">\n        <label [ngClass]=\"inline ? 'hc-form-field-label-inline' : 'hc-form-field-label'\" [class.hc-form-field-label-tight]=\"tight\" [attr.for]=\"_control._componentId\">\n            <ng-content select=\"hc-label\"></ng-content>\n            <span class=\"hc-required-marker\" aria-hidden=\"true\" *ngIf=\"_control._isRequired && !_control._isDisabled\">&nbsp;*</span>\n        </label>\n    </span>\n    <div class=\"hc-form-field-content-wrapper\">\n\n        <ng-content select=\"hc-label-ext\"></ng-content>\n\n        <div\n            *ngIf=\"hasInput\"\n            [class.hc-form-field-flex]=\"!inline\"\n            [class.hc-form-field-flex-inline]=\"inline\"\n            [class.hc-form-field-invalid]=\"_shouldShowErrorMessages()\"\n        >\n            <div class=\"hc-form-field-prefix\" *ngIf=\"_prefixChildren.length\">\n                <ng-content select=\"[hcPrefix]\"></ng-content>\n            </div>\n            <div #inputs class=\"hc-form-field-infix\" [class.hc-form-field-infix-tight]=\"tight\">\n                <ng-content select=\"[hcInput]\"></ng-content>\n            </div>\n            <div class=\"hc-form-field-suffix\" *ngIf=\"_suffixChildren.length\">\n                <ng-content select=\"[hcSuffix]\"></ng-content>\n            </div>\n        </div>\n\n        <div\n            *ngIf=\"!hasInput\"\n            [class.hc-form-field-non-input]=\"!inline\"\n            [class.hc-form-field-non-input-inline]=\"inline\"\n            [class.hc-form-field-non-input-inline-tight]=\"inline && tight\"\n            [class.hc-form-field-invalid]=\"_shouldShowErrorMessages()\"\n        >\n            <ng-content></ng-content>\n        </div>\n        <div class=\"hc-form-field-error-wrapper\" *ngIf=\"_shouldShowErrorMessages()\">\n            <hc-error *ngIf=\"_control._errorMessage && _control._errorMessage.length\">{{ _control._errorMessage }}</hc-error>\n            <ng-content select=\"hc-error\"></ng-content>\n        </div>\n    </div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".hc-form-field{position:relative;text-align:left;font-size:inherit;font-family:'Open Sans','Helvetica Neue',Helvetica,Arial,sans-serif;font-weight:400;line-height:1.5;-webkit-box-orient:inherit;-webkit-box-direction:inherit;flex-direction:inherit}.hc-form-field-wrapper{position:relative;padding-bottom:1.525em;-webkit-box-orient:inherit;-webkit-box-direction:inherit;flex-direction:inherit}.hc-form-field-wrapper-inline{display:-webkit-inline-box;display:inline-flex;width:100%;padding-bottom:inherit}.hc-form-field-wrapper-tight{padding-bottom:10px}.hc-form-field-content-wrapper{display:-webkit-box;display:flex;-webkit-box-orient:inherit;-webkit-box-direction:inherit;flex-direction:inherit;flex-wrap:wrap;width:100%}.hc-form-field-flex{display:-webkit-inline-box;display:inline-flex;-webkit-box-align:baseline;align-items:baseline;box-sizing:border-box;width:100%;border:1px solid #ccc;box-shadow:inset .5px .5px 0 #ccc,inset -.5px -.5px 0 #ccc;background:#fff}.hc-form-field-disabled .hc-form-field-flex{cursor:not-allowed;border:1px solid #e0e0e0;box-shadow:inset .5px .5px 0 #e0e0e0,inset -.5px -.5px 0 #e0e0e0;background-color:#f0f3f6;color:#adadad}.hc-form-field-flex-inline{display:-webkit-inline-box;display:inline-flex;-webkit-box-align:center;align-items:center;box-sizing:border-box;border:1px solid #ccc;box-shadow:inset .5px .5px 0 #ccc,inset -.5px -.5px 0 #ccc;background:#fff}.hc-form-field-disabled .hc-form-field-flex-inline{cursor:not-allowed;border:1px solid #e0e0e0;box-shadow:inset .5px .5px 0 #e0e0e0,inset -.5px -.5px 0 #e0e0e0;background-color:#f0f3f6;color:#adadad}.hc-form-field-invalid.hc-form-field-flex,.hc-form-field-invalid.hc-form-field-flex-inline{border:1px solid #cc2027;box-shadow:inset .5px .5px 0 #cc2027,inset -.5px -.5px 0 #cc2027}.hc-form-field-non-input{display:-webkit-inline-box;display:inline-flex;-webkit-box-orient:inherit;-webkit-box-direction:inherit;flex-direction:inherit;width:100%}.hc-form-field-non-input-inline{display:-webkit-inline-box;display:inline-flex;-webkit-box-flex:0;flex:0 0 100%;min-height:35px;-webkit-box-align:center;align-items:center}.hc-form-field-non-input-inline-tight{min-height:auto}.hc-form-field-prefix,.hc-form-field-suffix{white-space:nowrap;-webkit-box-flex:0;flex:none;position:relative}.hc-form-field-prefix .hc-icon,.hc-form-field-suffix .hc-icon{color:#a1a1a1;font-size:125%;line-height:1.5}.hc-form-field-wrapper-tight .hc-form-field-prefix .hc-icon,.hc-form-field-wrapper-tight .hc-form-field-suffix .hc-icon{line-height:1.3}.hc-form-field-disabled .hc-form-field-prefix,.hc-form-field-disabled .hc-form-field-suffix{opacity:.5}.hc-form-field-suffix{padding-right:7px}.hc-form-field-prefix{padding-left:7px}.hc-form-field-infix{display:block;position:relative;-webkit-box-flex:1;flex:auto;padding:.393em 10px}.hc-form-field-infix-tight{padding:.1965em 7px;font-size:.92857rem}.hc-form-field-label-wrapper{box-sizing:content-box;width:100%}.hc-form-field-label-wrapper .hc-icon{width:1em;height:1em;font-size:inherit;vertical-align:baseline}.hc-form-field-label-wrapper.hc-form-field-label-wrapper-tight{font-size:.92857rem}.hc-form-field-label-wrapper-inline{-webkit-box-flex:1;flex:1 0 auto;padding:7px 15px 0 0}.hc-form-field-label-wrapper-inline .hc-icon{width:1em;height:1em;font-size:inherit;vertical-align:baseline}.hc-form-field-label-wrapper-inline.hc-form-field-label-wrapper-tight{padding-top:0;line-height:1.9}.hc-form-field-label{display:block;width:100%;margin-bottom:3px;font:inherit;font-size:1rem;color:#333}.hc-form-field-label-extension{color:#6d6e70;font-size:.85714rem;margin-bottom:10px}.hc-form-field-label-inline{width:100%;font:inherit;font-size:1rem;color:#333}.hc-form-field-label-tight{font-size:.92857rem}.hc-form-field-error-wrapper{box-sizing:border-box;width:100%;overflow:hidden;margin-top:.53333em;font-size:75%}.hc-error{display:block;color:#cc2027}.hc-required-marker{color:#cc2027}@media all and (-ms-high-contrast:none){.hc-form-field-disabled-flex,.hc-form-field-flex,.hc-form-field-flex-inline,.hc-form-field-input-invalid{border-width:2px;box-shadow:none}}", ".hc-input{font:inherit;background:inherit;color:currentColor;border:none;outline:0;padding:0;margin:0;width:100%;max-width:100%;text-align:inherit}.hc-input[disabled]{cursor:not-allowed}.hc-input:-moz-ui-invalid{box-shadow:none}.hc-input::-ms-clear,.hc-input::-ms-reveal{display:none}.hc-input::-webkit-input-placeholder{color:#c0c5cc}.hc-input::-moz-placeholder{color:#c0c5cc}.hc-input::placeholder{color:#c0c5cc}.hc-input:-ms-input-placeholder{color:#c0c5cc}.hc-input::-ms-input-placeholder{color:#c0c5cc}textarea.hc-input{resize:vertical;overflow:auto;padding:2px 0;margin:-2px 0}"]
                }] }
    ];
    /** @nocollapse */
    HcFormFieldComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    HcFormFieldComponent.propDecorators = {
        _control: [{ type: ContentChild, args: [HcFormControlComponent, { static: false },] }],
        _controls: [{ type: ContentChildren, args: [HcFormControlComponent,] }],
        _errorChildren: [{ type: ContentChildren, args: [HcErrorComponent,] }],
        _prefixChildren: [{ type: ContentChildren, args: [HcPrefixDirective,] }],
        _suffixChildren: [{ type: ContentChildren, args: [HcSuffixDirective,] }],
        _inputChildren: [{ type: ContentChildren, args: [InputDirective,] }],
        _labelChildren: [{ type: ContentChildren, args: [HcLabelComponent,] }],
        _classHcFormFieldClass: [{ type: HostBinding, args: ['class.hc-form-field',] }],
        _disabledClass: [{ type: HostBinding, args: ['class.hc-form-field-disabled',] }],
        inline: [{ type: Input }],
        tight: [{ type: Input }]
    };
    return HcFormFieldComponent;
}());
export { HcFormFieldComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcFormFieldComponent.prototype._inline;
    /**
     * @type {?}
     * @private
     */
    HcFormFieldComponent.prototype._tight;
    /**
     * @type {?}
     * @private
     */
    HcFormFieldComponent.prototype.unsubscribe$;
    /** @type {?} */
    HcFormFieldComponent.prototype._control;
    /** @type {?} */
    HcFormFieldComponent.prototype._controls;
    /** @type {?} */
    HcFormFieldComponent.prototype._errorChildren;
    /** @type {?} */
    HcFormFieldComponent.prototype._prefixChildren;
    /** @type {?} */
    HcFormFieldComponent.prototype._suffixChildren;
    /** @type {?} */
    HcFormFieldComponent.prototype._inputChildren;
    /** @type {?} */
    HcFormFieldComponent.prototype._labelChildren;
    /** @type {?} */
    HcFormFieldComponent.prototype._classHcFormFieldClass;
    /**
     * @type {?}
     * @private
     */
    HcFormFieldComponent.prototype._elementRef;
}
