// @healthcatalyst/cashmere/esm5/lib/form-field/hc-form.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/form-field/hc-form.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ContentChildren, QueryList } from '@angular/core';
import { parseBooleanAttribute } from '../util';
import { HcFormFieldComponent } from './hc-form-field.component';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
/**
 * `hcForm` directive that allows settings to be applied to all included HcFormFields
 */
var HcFormDirective = /** @class */ (function () {
    function HcFormDirective() {
        this._tight = false;
        this.unsubscribe$ = new Subject();
    }
    Object.defineProperty(HcFormDirective.prototype, "tight", {
        /** Set the tight parameter on all enclosed HcFormFields. *Defaults to `false`.*  */
        get: /**
         * Set the tight parameter on all enclosed HcFormFields. *Defaults to `false`.*
         * @return {?}
         */
        function () {
            return this._tight;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._tight = parseBooleanAttribute(value);
            this._updateTightFields();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HcFormDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._updateTightFields();
        // Pass the tight setting to any FormFields added dynamically
        this._formFields.changes.pipe(takeUntil(this.unsubscribe$)).subscribe((/**
         * @return {?}
         */
        function () { return _this._updateTightFields(); }));
    };
    /**
     * @return {?}
     */
    HcFormDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    };
    /**
     * @return {?}
     */
    HcFormDirective.prototype._updateTightFields = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._formFields) {
            this._formFields.forEach((/**
             * @param {?} field
             * @return {?}
             */
            function (field) {
                field.tight = _this._tight;
            }));
        }
    };
    HcFormDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hcForm]'
                },] }
    ];
    HcFormDirective.propDecorators = {
        _formFields: [{ type: ContentChildren, args: [HcFormFieldComponent,] }],
        tight: [{ type: Input }]
    };
    return HcFormDirective;
}());
export { HcFormDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcFormDirective.prototype._tight;
    /**
     * @type {?}
     * @private
     */
    HcFormDirective.prototype.unsubscribe$;
    /** @type {?} */
    HcFormDirective.prototype._formFields;
}
