// @healthcatalyst/cashmere/esm5/lib/tabs/tab/tab.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/tabs/tab/tab.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChildren, QueryList, Output, ViewEncapsulation } from '@angular/core';
import { EventEmitter, TemplateRef, ViewChild } from '@angular/core';
import { HcTabTitleComponent } from './tab-title.component';
var TabComponent = /** @class */ (function () {
    function TabComponent() {
        /**
         * Plain text title of the tab; for HTML support include a `hc-tab-title` element
         */
        this.tabTitle = '';
        /**
         * Emits when this tab is selected; use instead of `(click)` for click binding
         */
        this.tabClick = new EventEmitter();
        this._active = false;
        this._tight = false;
    }
    /**
     * @return {?}
     */
    TabComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (this._tabTitle) {
            this._htmlTitle = this._tabTitle.first;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    TabComponent.prototype.tabClickHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // Prevent a tab anchor click from also calling the router on the host element
        event.preventDefault();
        event.stopPropagation();
        this.tabClick.emit();
    };
    TabComponent.decorators = [
        { type: Component, args: [{
                    template: "<ng-container *ngIf=\"_direction\">\n    <a\n        *ngIf=\"routerLink; else tabWithoutRouting\"\n        class=\"hc-tab-{{ _direction }} hc-text-ellipsis\"\n        [class.hc-tab-tight]=\"_tight\"\n        [class.active]=\"_active\"\n        [routerLink]=\"routerLink\"\n        routerLinkActive=\"active\"\n        [queryParams]=\"queryParams\"\n        [routerLinkActiveOptions]=\"exactRouteMatch ? {exact: true} : {}\"\n        (click)=\"tabClickHandler($event)\"\n    >\n        <ng-container *ngIf=\"_htmlTitle\" [ngTemplateOutlet]=\"_htmlTitle.tabTitle\"></ng-container>\n        {{ tabTitle }}\n    </a>\n</ng-container>\n<ng-template #tabWithoutRouting>\n    <a\n        class=\"hc-tab-{{ _direction }} hc-text-ellipsis\"\n        [class.hc-tab-tight]=\"_tight\"\n        [class.active]=\"_active\"\n        (click)=\"tabClickHandler($event)\"\n    >\n        <ng-container *ngIf=\"_htmlTitle\" [ngTemplateOutlet]=\"_htmlTitle.tabTitle\"></ng-container>\n        {{ tabTitle }}\n    </a>\n</ng-template>\n<ng-template #tabContent>\n    <ng-content></ng-content>\n</ng-template>\n",
                    selector: "hc-tab",
                    // tslint:disable-next-line: no-host-metadata-property
                    host: { class: 'hc-tab' },
                    encapsulation: ViewEncapsulation.None,
                    styles: [".hc-tab{min-width:0}.hc-tab-vertical{background-color:#fff;color:#333;display:block;padding:20px;text-decoration:none}.hc-tab-vertical:hover{cursor:pointer}.hc-tab-vertical.active{background-color:#f0f3f6;border-left:8px solid #00aeff;color:#00aeff;font-weight:600;padding-left:12px}.hc-tab-vertical.hc-tab-tight{padding:10px 20px}.hc-tab-horizontal{background-color:inherit;color:#333;display:block;font-size:14px;min-width:100px;padding:10px 20px;text-align:center;text-decoration:none}.hc-tab-horizontal:hover{cursor:pointer}.hc-tab-horizontal.active{border-bottom:4px solid #00aeff;color:#333;font-weight:700}.hc-tab-horizontal.hc-tab-tight{padding:10px}"]
                }] }
    ];
    TabComponent.propDecorators = {
        tabTitle: [{ type: Input }],
        routerLink: [{ type: Input }],
        queryParams: [{ type: Input }],
        exactRouteMatch: [{ type: Input }],
        tabClick: [{ type: Output }],
        tabContent: [{ type: ViewChild, args: ['tabContent', { static: false },] }],
        _tabTitle: [{ type: ContentChildren, args: [HcTabTitleComponent,] }]
    };
    return TabComponent;
}());
export { TabComponent };
if (false) {
    /**
     * Plain text title of the tab; for HTML support include a `hc-tab-title` element
     * @type {?}
     */
    TabComponent.prototype.tabTitle;
    /**
     * Router path that the tab routes to. If one tab uses the routerLink in a tab set, all must use the router link.
     * Can be specified as '/path/2' or ['path', '2']
     * @type {?}
     */
    TabComponent.prototype.routerLink;
    /**
     * Add queryParams to the routerLink
     * @type {?}
     */
    TabComponent.prototype.queryParams;
    /**
     * If set to true, [routerLinkActiveOptions]="{exact: true}" is added
     * @type {?}
     */
    TabComponent.prototype.exactRouteMatch;
    /**
     * Emits when this tab is selected; use instead of `(click)` for click binding
     * @type {?}
     */
    TabComponent.prototype.tabClick;
    /**
     * The template to be used when this tab is selected. Defaults to the content of this tab component.
     * Not used when the tab set uses routing.
     * @type {?}
     */
    TabComponent.prototype.tabContent;
    /** @type {?} */
    TabComponent.prototype._direction;
    /** @type {?} */
    TabComponent.prototype._active;
    /** @type {?} */
    TabComponent.prototype._tight;
    /** @type {?} */
    TabComponent.prototype._htmlTitle;
    /** @type {?} */
    TabComponent.prototype._tabTitle;
}
