// @healthcatalyst/cashmere/esm5/lib/app-switcher/app-switcher.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/app-switcher/app-switcher.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { AppSwitcherComponent } from './app-switcher.component';
import { PopModule } from '../pop/popover.module';
import { ProgressIndicatorsModule } from '../progress-indicators/progress-indicators.module';
import { AppSwitcherService } from './app-switcher.service';
import { WorkTrackerService } from '../shared/work-tracker.service';
import { PipesModule } from '../pipes/pipes.module';
import { APP_SWITCHER_CONFIG, APP_SWITCHER_SERVICE } from './app-switcher-interfaces';
import { AppSwitcherLinksComponent } from './app-switcher-application-link/app-switcher-links.component';
import { IconModule } from '../icon/icon.module';
var AppSwitcherModule = /** @class */ (function () {
    function AppSwitcherModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    AppSwitcherModule.forRoot = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: AppSwitcherModule,
            providers: [
                {
                    provide: APP_SWITCHER_CONFIG,
                    useValue: config
                }
            ]
        };
    };
    AppSwitcherModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, PopModule, HttpClientModule, PipesModule, IconModule, ProgressIndicatorsModule],
                    declarations: [AppSwitcherComponent, AppSwitcherLinksComponent],
                    exports: [AppSwitcherComponent, AppSwitcherLinksComponent],
                    entryComponents: [AppSwitcherComponent],
                    providers: [
                        {
                            provide: APP_SWITCHER_SERVICE,
                            useClass: AppSwitcherService
                        },
                        WorkTrackerService
                    ]
                },] }
    ];
    return AppSwitcherModule;
}());
export { AppSwitcherModule };
