// @healthcatalyst/cashmere/esm2015/lib/date-range/date-range/date-range.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/date-range/date-range/date-range.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Output, EventEmitter, Input, ElementRef, Directive, HostListener } from '@angular/core';
import { DatePipe } from '@angular/common';
import { CalendarOverlayService } from '../services/calendar-overlay.service';
import { ConfigStoreService } from '../services/config-store.service';
export class DateRangeDirective {
    /**
     * @param {?} _elementRef
     * @param {?} calendarOverlayService
     * @param {?} configStoreService
     */
    constructor(_elementRef, calendarOverlayService, configStoreService) {
        this._elementRef = _elementRef;
        this.calendarOverlayService = calendarOverlayService;
        this.configStoreService = configStoreService;
        /**
         * Emits when date range is been changed.
         */
        this.selectedDateRangeChanged = new EventEmitter();
        /**
         * Emits either a numerical index for the selected preset, or a `DateRange` if the selected value is not a preset
         */
        this.selectedPresetChanged = new EventEmitter();
        configStoreService.rangeUpdate$.subscribe((/**
         * @param {?} daterange
         * @return {?}
         */
        (daterange) => {
            this.selectedDateRangeChanged.emit(daterange);
        }));
        configStoreService.presetUpdate$.subscribe((/**
         * @param {?} preset
         * @return {?}
         */
        (preset) => {
            this.selectedPresetChanged.emit(preset);
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._overlayRef) {
            this._overlayRef.detach();
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['options']) {
            /** @type {?} */
            const options = changes['options'].currentValue;
            this.configStoreService.updateDateRangeOptions(options);
        }
        if (changes['selectedDate']) {
            /** @type {?} */
            const selectedDate = changes['selectedDate'].currentValue;
            if (typeof selectedDate === 'number') {
                this.configStoreService.updatePreset(selectedDate);
            }
            else {
                this.configStoreService.updateRange(selectedDate);
            }
        }
    }
    /**
     * @return {?}
     */
    _onClick() {
        this._overlayRef = this.calendarOverlayService.open(this._elementRef);
    }
}
DateRangeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hcDateRange]',
                providers: [CalendarOverlayService, ConfigStoreService, DatePipe]
            },] }
];
/** @nocollapse */
DateRangeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: CalendarOverlayService },
    { type: ConfigStoreService }
];
DateRangeDirective.propDecorators = {
    selectedDateRangeChanged: [{ type: Output }],
    selectedDate: [{ type: Input }],
    selectedPresetChanged: [{ type: Output }],
    options: [{ type: Input }],
    _onClick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /**
     * Emits when date range is been changed.
     * @type {?}
     */
    DateRangeDirective.prototype.selectedDateRangeChanged;
    /**
     * Sets the selected date range. Accepts either a `DateRange` or a numerical index for preset.
     * @type {?}
     */
    DateRangeDirective.prototype.selectedDate;
    /**
     * Emits either a numerical index for the selected preset, or a `DateRange` if the selected value is not a preset
     * @type {?}
     */
    DateRangeDirective.prototype.selectedPresetChanged;
    /**
     * Configuration to setup behavior of component.
     * @type {?}
     */
    DateRangeDirective.prototype.options;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype._overlayRef;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    DateRangeDirective.prototype.calendarOverlayService;
    /** @type {?} */
    DateRangeDirective.prototype.configStoreService;
}
