// @healthcatalyst/cashmere/esm2015/lib/scroll-nav/nav/scroll-nav.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/scroll-nav/nav/scroll-nav.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewEncapsulation, QueryList, ContentChildren } from '@angular/core';
import { ScrollNavLinkDirective } from './scroll-nav-link.directive';
/**
 * Container for scroll navigation links.
 */
export class HcScrollNavComponent {
    /**
     * @param {?} _elementRef
     */
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        this.ACTIVE_CLASS = 'hc-scroll-nav-link-active';
    }
    /**
     * @return {?}
     */
    get _links() {
        return this.linkList.toArray().map((/**
         * @param {?} e
         * @return {?}
         */
        e => e._el.nativeElement));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this._links && this._links[0]) {
            this._links[0].classList.add(this.ACTIVE_CLASS);
        }
    }
    /**
     * @param {?} id
     * @return {?}
     */
    _setActiveClassById(id) {
        /** @type {?} */
        const link = this.linkList.find((/**
         * @param {?} e
         * @return {?}
         */
        e => e.hcScrollLink === id));
        if (!link) {
            throw new Error(`Failed to mark active class. Could not find the element with the data target for id: ${id}.`);
        }
        this.setActiveClass(link._el.nativeElement);
    }
    /**
     * @private
     * @param {?} element
     * @return {?}
     */
    setActiveClass(element) {
        this._links.forEach((/**
         * @param {?} e
         * @return {?}
         */
        e => {
            e.classList.remove(this.ACTIVE_CLASS);
        }));
        element.classList.add(this.ACTIVE_CLASS);
    }
}
HcScrollNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'hc-scroll-nav',
                encapsulation: ViewEncapsulation.None,
                template: "<div class=\"hc-scroll-nav-links-container\">\n  <ng-content></ng-content>\n</div>",
                styles: [".hc-scroll-nav-links-container{margin-right:30px}.hc-scroll-nav-links-container ul{list-style:none}.hc-scroll-nav-link{font-size:1rem;background-color:transparent;color:#333;cursor:pointer;font-weight:500;padding:10px 15px}.hc-scroll-nav-link:hover{background-color:rgba(215,221,228,.25)}.hc-scroll-nav-link:active{background-color:rgba(215,221,228,.5)}.hc-scroll-nav-link-active,.hc-scroll-nav-link-active:hover{background-color:#00aeff;color:#fff;font-weight:500}"]
            }] }
];
/** @nocollapse */
HcScrollNavComponent.ctorParameters = () => [
    { type: ElementRef }
];
HcScrollNavComponent.propDecorators = {
    linkList: [{ type: ContentChildren, args: [ScrollNavLinkDirective,] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    HcScrollNavComponent.prototype.linkList;
    /**
     * @type {?}
     * @private
     */
    HcScrollNavComponent.prototype.ACTIVE_CLASS;
    /** @type {?} */
    HcScrollNavComponent.prototype._elementRef;
}
